// @nevermined-io/payments/dist/a2a/paymentsRequestHandler.js
import { A2AError, DefaultRequestHandler, ExecutionEventQueue, ResultManager, } from '@a2a-js/sdk/server';
import { v4 as uuidv4 } from 'uuid';
import { PaymentsError } from '../common/payments.error.js';
import { isValidScheme } from '../common/types.js';
import { decodeAccessToken } from '../utils.js';
import { buildPaymentRequired } from '../x402/facilitator-api.js';
const terminalStates = ['completed', 'failed', 'canceled', 'rejected'];
/**
 * PaymentsRequestHandler extends DefaultRequestHandler to add payments validation and burning.
 * It validates credits before executing a task and burns credits after successful execution.
 * It also sends push notifications when a task reaches a terminal state.
 * @param options - Handler options, including asyncExecution, defaultBatch, and defaultMarginPercent
 */
export class PaymentsRequestHandler extends DefaultRequestHandler {
    /**
     * Store HTTP context temporarily by messageId (used in middleware when taskId is not yet available).
     * @param messageId - The messageId from the incoming message
     * @param ctx - The HTTP context (bearerToken, url, method)
     */
    setHttpRequestContextForMessage(messageId, ctx) {
        this.httpContextByMessageId.set(messageId, ctx);
    }
    /**
     * Store HTTP context by taskId (used in middleware when taskId is available).
     * @param taskId - The taskId of the task
     * @param ctx - The HTTP context (bearerToken, url, method)
     */
    setHttpRequestContextForTask(taskId, ctx) {
        this.httpContextByTaskId.set(taskId, ctx);
    }
    /**
     * @param agentCard - The agent card
     * @param taskStore - The task store
     * @param agentExecutor - The business logic executor
     * @param paymentsService - The payments service for validation and burning
     * @param eventBusManager - The event bus manager (optional)
     * @param options - Handler options (asyncExecution, defaultBatch, defaultMarginPercent)
     */
    constructor(agentCard, taskStore, agentExecutor, paymentsService, eventBusManager, options) {
        super(agentCard, taskStore, agentExecutor, eventBusManager);
        this.httpContextByTaskId = new Map();
        this.httpContextByMessageId = new Map();
        this.paymentsService = paymentsService;
        this.asyncExecution = options?.asyncExecution ?? false;
        this.defaultBatch = options?.defaultBatch ?? false;
        this.defaultMarginPercent = options?.defaultMarginPercent;
    }
    /**
     * Retrieve the HTTP context for a given taskId.
     * @param taskId - The taskId of the task
     * @returns The HTTP context (bearerToken, url, method) or undefined
     */
    getHttpRequestContextForTask(taskId) {
        return this.httpContextByTaskId.get(taskId);
    }
    /**
     * Retrieve the HTTP context for a given messageId.
     * @param messageId - The messageId of the message
     * @returns The HTTP context (bearerToken, url, method) or undefined
     */
    getHttpRequestContextForMessage(messageId) {
        return this.httpContextByMessageId.get(messageId);
    }
    /**
     * Deletes the HTTP context associated with a taskId.
     * @param taskId - The taskId to delete context for
     */
    deleteHttpRequestContextForTask(taskId) {
        this.httpContextByTaskId.delete(taskId);
    }
    /**
     * Get the handler options (defaultBatch, defaultMarginPercent).
     * Used by middleware to determine default redemption behavior.
     * @returns The handler options
     */
    getHandlerOptions() {
        return {
            asyncExecution: this.asyncExecution,
            defaultBatch: this.defaultBatch,
            defaultMarginPercent: this.defaultMarginPercent,
        };
    }
    /**
     * Validates a request using the x402 payments service.
     * This method is used by the middleware to validate credits before processing requests.
     *
     * @param bearerToken - The bearer token for authentication
     * @param endpoint - Optional endpoint URL being requested
     * @param httpVerb - Optional HTTP method being used
     * @returns Promise resolving to the validation result
     */
    async validateRequest(bearerToken, endpoint, httpVerb) {
        let planId;
        const agentCard = await this.getAgentCard();
        const paymentExtension = agentCard.capabilities?.extensions?.find((ext) => ext.uri === 'urn:nevermined:payment');
        if (paymentExtension) {
            planId = paymentExtension.params?.planId;
        }
        const decodedAccessToken = decodeAccessToken(bearerToken);
        if (!decodedAccessToken) {
            throw PaymentsError.unauthorized('Invalid access token.');
        }
        if (!planId) {
            throw PaymentsError.unauthorized('Plan ID not found in agent card.');
        }
        // Extract subscriberAddress from token (payload.authorization.from per x402 spec)
        const subscriberAddress = decodedAccessToken.payload?.authorization?.from;
        if (!subscriberAddress) {
            throw PaymentsError.unauthorized('Cannot determine subscriberAddress from token (expected payload.authorization.from)');
        }
        const agentId = paymentExtension?.params?.agentId;
        const scheme = isValidScheme(decodedAccessToken.accepted?.scheme) ? decodedAccessToken.accepted.scheme : 'nvm:erc4337';
        const paymentRequired = buildPaymentRequired(planId, {
            endpoint: endpoint || '',
            agentId,
            httpVerb,
            scheme,
            environment: this.paymentsService.getEnvironmentName(),
        });
        const result = await this.paymentsService.facilitator.verifyPermissions({
            paymentRequired,
            x402AccessToken: bearerToken,
            maxAmount: 1n,
        });
        if (!result.isValid) {
            throw PaymentsError.unauthorized('Permission verification failed.');
        }
        return {
            success: true,
            planId,
            subscriberAddress,
            balance: { isSubscriber: true },
        };
    }
    /**
     * Gets redemption configuration for a task based on AgentCard and handler defaults.
     * The configuration is determined by the server, not by client metadata.
     *
     * @returns The redemption configuration
     */
    async getRedemptionConfig() {
        const agentCard = await this.getAgentCard();
        const paymentExtension = agentCard.capabilities?.extensions?.find((ext) => ext.uri === 'urn:nevermined:payment');
        const agentConfig = paymentExtension?.params?.redemptionConfig || {};
        return {
            useBatch: agentConfig.useBatch ?? this.defaultBatch ?? false,
            useMargin: agentConfig.useMargin ?? false,
            marginPercent: agentConfig.marginPercent ?? this.defaultMarginPercent,
        };
    }
    /**
     * Determines the appropriate redemption method based on server configuration.
     *
     * @param bearerToken - The bearer token for authentication
     * @param creditsUsed - The number of credits to burn
     * @param httpContext - Optional HTTP context with endpoint and method information
     * @returns Promise resolving to the redemption result
     */
    async executeRedemption(bearerToken, creditsUsed, httpContext) {
        const decodedAccessToken = decodeAccessToken(bearerToken);
        if (!decodedAccessToken) {
            throw PaymentsError.unauthorized('Invalid access token.');
        }
        let planId;
        let agentId;
        const agentCard = await this.getAgentCard();
        const paymentExtension = agentCard.capabilities?.extensions?.find((ext) => ext.uri === 'urn:nevermined:payment');
        if (paymentExtension) {
            planId = paymentExtension.params?.planId;
            agentId = paymentExtension.params?.agentId;
        }
        if (!planId) {
            throw PaymentsError.unauthorized('Plan ID not found in agent card.');
        }
        const scheme = isValidScheme(decodedAccessToken.accepted?.scheme) ? decodedAccessToken.accepted.scheme : 'nvm:erc4337';
        // Build paymentRequired using the helper
        const paymentRequired = buildPaymentRequired(planId, {
            endpoint: httpContext?.urlRequested,
            agentId,
            httpVerb: httpContext?.httpMethodRequested,
            scheme,
            environment: this.paymentsService.getEnvironmentName(),
        });
        return await this.paymentsService.facilitator.settlePermissions({
            paymentRequired,
            x402AccessToken: bearerToken,
            maxAmount: BigInt(creditsUsed),
        });
    }
    /**
     * Creates PaymentsRequestContext from message parameters.
     * This method handles HTTP context retrieval, validation, and context creation.
     * @param params - Message send parameters
     * @param isStreaming - Whether this is for streaming (affects createRequestContext call)
     * @returns Object containing PaymentsRequestContext and related data
     */
    async createPaymentsRequestContext(params, isStreaming = false) {
        // Validate required parameters before any processing
        const missingParam = !params.message
            ? 'message'
            : !params.message.messageId
                ? 'message.messageId'
                : null;
        if (missingParam) {
            throw A2AError.invalidParams(`${missingParam} is required.`);
        }
        // 1. Get HTTP context for the task or message
        let taskId = params.message.taskId;
        let httpContext;
        if (taskId) {
            httpContext = this.getHttpRequestContextForTask(taskId);
        }
        else {
            const messageId = params.message.messageId;
            httpContext = this.getHttpRequestContextForMessage(messageId);
        }
        if (!httpContext) {
            throw A2AError.internalError('HTTP context not found for task or message.');
        }
        // 2. Extract bearer token and validate presence of required fields
        const { bearerToken, validation } = httpContext;
        if (!bearerToken) {
            throw PaymentsError.unauthorized('Missing bearer token for payment validation.');
        }
        // 3. Validate credits before executing the task
        const agentCard = await this.getAgentCard();
        const agentId = agentCard.capabilities?.extensions?.find((ext) => ext.uri === 'urn:nevermined:payment')?.params?.agentId;
        if (!agentId) {
            throw A2AError.internalError('Agent ID not found in payment extension.');
        }
        // 4. Generate taskId if not present and migrate HTTP context
        const incomingMessage = params.message;
        if (!taskId) {
            taskId = uuidv4();
            this.migrateHttpRequestContextFromMessageToTask(params.message.messageId, taskId);
        }
        // 5. Instantiate ResultManager and eventBus
        const resultManager = new ResultManager(this.getTaskStore());
        resultManager.setContext(incomingMessage);
        const requestContext = await this.callCreateRequestContext(incomingMessage, taskId, isStreaming);
        const finalMessageForAgent = requestContext.userMessage;
        const eventBus = this.getEventBusManager().createOrGetByTaskId(taskId);
        const eventQueue = new ExecutionEventQueue(eventBus);
        // 6. Create PaymentsRequestContext
        const authResult = {
            requestId: validation.agentRequestId,
            token: bearerToken,
            agentId: validation.agentId,
            agentRequest: validation,
        };
        const agentRequestContext = {
            authResult,
            httpContext,
            paymentsService: this.paymentsService,
        };
        // 7. Create extended request context with agent request data
        const paymentsRequestContext = {
            ...requestContext,
            payments: agentRequestContext,
        };
        return {
            paymentsRequestContext,
            taskId,
            httpContext,
            bearerToken,
            validation,
            requestContext,
            finalMessageForAgent,
            eventBus,
            eventQueue,
            resultManager,
        };
    }
    /**
     * Processes streaming events with finalization (credits burning and push notifications).
     * Similar to processEventsWithFinalization but yields events for streaming.
     * @param taskId - The task ID
     * @param resultManager - The result manager
     * @param eventQueue - The event queue
     * @param bearerToken - The bearer token
     * @param validation - The validation result
     * @returns Async generator yielding processed events
     */
    async *processStreamingEventsWithFinalization(taskId, resultManager, eventQueue, bearerToken) {
        try {
            for await (const event of eventQueue.events()) {
                await resultManager.processEvent(event);
                yield event;
                if (event.kind === 'status-update' &&
                    event.final &&
                    event?.metadata?.creditsUsed !== undefined &&
                    event?.metadata?.creditsUsed !== null &&
                    bearerToken &&
                    (typeof event.metadata.creditsUsed === 'string' ||
                        typeof event.metadata.creditsUsed === 'number' ||
                        typeof event.metadata.creditsUsed === 'bigint')) {
                    try {
                        // Get redemption configuration from server (not from client metadata)
                        const redemptionConfig = await this.getRedemptionConfig();
                        if (!redemptionConfig.useBatch) {
                            // Get HTTP context if available
                            const httpContext = this.getHttpRequestContextForTask(event.taskId);
                            // Execute redemption with server configuration for non-batch requests
                            const response = await this.executeRedemption(bearerToken, BigInt(event.metadata.creditsUsed), httpContext);
                            // Update event metadata with response data
                            if (response && event.metadata) {
                                event.metadata.txHash = response.txHash;
                                event.metadata.creditsCharged = response.amountOfCredits
                                    ? Number(response.amountOfCredits)
                                    : event.metadata.creditsUsed;
                            }
                        }
                    }
                    catch (err) {
                        // Do nothing
                    }
                }
                // Handle push notification
                if (event.kind === 'status-update' &&
                    event.final &&
                    event.status?.state &&
                    terminalStates.includes(event.status.state)) {
                    try {
                        const taskPushNotificationConfig = await this.getTaskPushNotificationConfig({
                            id: event.taskId,
                        });
                        if (taskPushNotificationConfig) {
                            await this.sendPushNotification(event.taskId, event.status.state, taskPushNotificationConfig.pushNotificationConfig, {
                                contextId: event.contextId,
                            });
                        }
                    }
                    catch (err) {
                        // Do nothing
                    }
                }
            }
        }
        finally {
            // Cleanup when the stream is fully consumed or breaks
            //this.getEventBusManager().cleanupByTaskId(taskId)
        }
    }
    /**
     * Processes all events, calling handleTaskFinalization when a terminal status-update event is received.
     * In async mode, it can be launched in background.
     */
    async processEventsWithFinalization(taskId, resultManager, eventQueue, bearerToken, validation, options) {
        let firstResultSent = false;
        try {
            for await (const event of eventQueue.events()) {
                // Handle redemption before processing the event
                if (event.kind === 'status-update' &&
                    event.final &&
                    terminalStates.includes(event.status?.state)) {
                    // Get HTTP context if available
                    const httpContext = this.getHttpRequestContextForTask(event.taskId);
                    await this.handleTaskFinalization(resultManager, event, bearerToken, httpContext);
                }
                await resultManager.processEvent(event);
                if (options?.firstResultResolver && !firstResultSent) {
                    if (event.kind === 'message' || event.kind === 'task') {
                        options.firstResultResolver(event);
                        firstResultSent = true;
                    }
                }
            }
            if (options?.firstResultRejector && !firstResultSent) {
                options.firstResultRejector(A2AError.internalError('Execution finished before a message or task was produced.'));
            }
        }
        catch (error) {
            if (options?.firstResultRejector && !firstResultSent) {
                options.firstResultRejector(error);
            }
            throw error;
        }
        finally {
            this.getEventBusManager().cleanupByTaskId(taskId);
        }
    }
    /**
     * Sends a message, validating credits before execution and burning credits after.
     * Also sends a push notification if the task reaches a terminal state.
     * This method overrides the parent implementation to allow eventBus subscription before agent execution.
     * @param params - Message send parameters
     * @returns The resulting message or task
     */
    async sendMessage(params) {
        // Create PaymentsRequestContext and related data
        const { paymentsRequestContext, taskId, bearerToken, validation, requestContext, finalMessageForAgent, eventBus, eventQueue, resultManager, } = await this.createPaymentsRequestContext(params, false);
        this.agentExecutor
            .execute(paymentsRequestContext, eventBus)
            .catch((err) => {
            const errorTask = {
                id: requestContext.task?.id || uuidv4(),
                contextId: finalMessageForAgent.contextId || uuidv4(),
                status: {
                    state: 'failed',
                    message: {
                        kind: 'message',
                        role: 'agent',
                        messageId: uuidv4(),
                        parts: [{ kind: 'text', text: `Agent execution error: ${err.message}` }],
                        taskId: requestContext.task?.id,
                        contextId: finalMessageForAgent.contextId,
                    },
                    timestamp: new Date().toISOString(),
                },
                history: requestContext.task?.history ? [...requestContext.task.history] : [],
                kind: 'task',
            };
            if (finalMessageForAgent) {
                if (!errorTask.history?.find((m) => m.messageId === finalMessageForAgent.messageId)) {
                    errorTask.history?.push(finalMessageForAgent);
                }
            }
            eventBus.publish(errorTask);
            eventBus.publish({
                kind: 'status-update',
                taskId: errorTask.id,
                contextId: errorTask.contextId,
                status: errorTask.status,
                final: true,
            });
            eventBus.finished();
        });
        // Determine if execution should be blocking based on client request
        // The blocking parameter comes from params.configuration.blocking
        const isBlocking = params.configuration?.blocking !== false; // Default to blocking if not specified
        if (isBlocking) {
            await this.processEventsWithFinalization(taskId, resultManager, eventQueue, bearerToken, validation);
            const finalResult = resultManager.getFinalResult();
            if (!finalResult) {
                throw A2AError.internalError('Agent execution finished without a result, and no task context found.');
            }
            return finalResult;
        }
        else {
            // Non-blocking execution - return immediately with first result
            if (!taskId) {
                throw A2AError.internalError('Task ID is required for non-blocking execution.');
            }
            const validTaskId = taskId;
            return new Promise((resolve, reject) => {
                this.processEventsWithFinalization(validTaskId, resultManager, eventQueue, bearerToken, validation, {
                    firstResultResolver: resolve,
                    firstResultRejector: reject,
                });
            });
        }
    }
    /**
     * Handles credits burning and push notification when a task reaches a terminal state.
     * This is called asynchronously from the eventBus listener.
     * Supports batch and margin-based redemptions based on server configuration.
     * @param resultManager - The result manager
     * @param event - The status-update event with final state
     * @param bearerToken - The bearer token for payment validation
     * @param validation - The validation result from the request
     */
    async handleTaskFinalization(resultManager, event, bearerToken, httpContext) {
        const creditsToBurn = event.metadata?.creditsUsed;
        if (creditsToBurn !== undefined &&
            creditsToBurn !== null &&
            bearerToken &&
            (typeof creditsToBurn === 'string' ||
                typeof creditsToBurn === 'number' ||
                typeof creditsToBurn === 'bigint')) {
            try {
                // Get redemption configuration from server (not from client metadata)
                const redemptionConfig = await this.getRedemptionConfig();
                if (!redemptionConfig.useBatch) {
                    // Execute redemption with server configuration for non-batch requests
                    const response = await this.executeRedemption(bearerToken, BigInt(creditsToBurn), httpContext);
                    // Update event metadata with redemption results
                    event.metadata = {
                        ...event.metadata,
                        txHash: response.txHash,
                        // Store the actual credits charged (especially important for margin-based)
                        creditsCharged: response.amountOfCredits
                            ? Number(response.amountOfCredits)
                            : creditsToBurn,
                    };
                }
                // Always update task metadata and process the task
                const task = resultManager.getCurrentTask();
                if (task) {
                    // Update task metadata with current event metadata (from executor or redemption)
                    task.metadata = {
                        ...task.metadata,
                        ...event.metadata,
                    };
                    await resultManager.processEvent(task);
                    // Delete http context associated with the task
                    this.deleteHttpRequestContextForTask(event.taskId);
                }
            }
            catch (err) {
                // Do nothing
            }
        }
        try {
            const taskPushNotificationConfig = await this.getTaskPushNotificationConfig({
                id: event.taskId,
            });
            if (taskPushNotificationConfig) {
                await this.sendPushNotification(event.taskId, event.status.state, taskPushNotificationConfig.pushNotificationConfig, { contextId: event.contextId });
            }
        }
        catch (err) {
            // Do nothing
        }
    }
    /**
     * Streams messages and events for a task, with payments validation.
     * Also sends a push notification if a terminal status-update event is emitted.
     * @param params - Message send parameters
     * @returns Async generator of events
     */
    async *sendMessageStream(params) {
        // Create PaymentsRequestContext and related data
        const { paymentsRequestContext, taskId, bearerToken, requestContext, finalMessageForAgent, eventBus, eventQueue, resultManager, } = await this.createPaymentsRequestContext(params, true);
        this.agentExecutor
            .execute(paymentsRequestContext, eventBus)
            .catch((err) => {
            console.error(`Agent execution failed for stream message ${finalMessageForAgent.messageId}:`, err);
            const contextId = finalMessageForAgent.contextId || uuidv4();
            const errorTaskStatus = {
                kind: 'status-update',
                taskId: requestContext.task?.id || uuidv4(),
                contextId,
                status: {
                    state: 'failed',
                    message: {
                        kind: 'message',
                        role: 'agent',
                        messageId: uuidv4(),
                        parts: [{ kind: 'text', text: `Agent execution error: ${err.message}` }],
                        taskId: requestContext.task?.id,
                        contextId,
                    },
                    timestamp: new Date().toISOString(),
                },
                final: true,
            };
            eventBus.publish(errorTaskStatus);
        });
        // Process streaming events with finalization
        yield* this.processStreamingEventsWithFinalization(taskId, resultManager, eventQueue, bearerToken);
    }
    /**
     * Sends a push notification when a task reaches a terminal state.
     * @param taskId - The task ID
     * @param state - The terminal state
     * @param pushNotificationConfig - The push notification configuration
     * @param payload - Additional payload to include in the notification
     */
    async sendPushNotification(taskId, state, pushNotificationConfig, payload = {}) {
        const headers = {
            'Content-Type': 'application/json',
        };
        if (pushNotificationConfig.authentication) {
            if (pushNotificationConfig.authentication.schemes.includes('basic')) {
                const basic = Buffer.from(`${pushNotificationConfig.authentication.credentials}`).toString('base64');
                headers['Authorization'] = `Basic ${basic}`;
            }
            else if (pushNotificationConfig.authentication.schemes.includes('bearer')) {
                headers['Authorization'] = `Bearer ${pushNotificationConfig.authentication.credentials}`;
            }
            else if (pushNotificationConfig.authentication.schemes.includes('custom')) {
                Object.assign(headers, pushNotificationConfig.authentication.credentials);
            }
        }
        const body = JSON.stringify({
            taskId,
            state,
            payload,
        });
        const response = await fetch(pushNotificationConfig.url, {
            method: 'POST',
            headers,
            body,
        });
        if (!response.ok) {
            throw A2AError.internalError('Failed to send push notification.');
        }
    }
    /**
     * Migrates the HTTP context from a messageId to a taskId and deletes the temporary messageId context.
     * @param messageId - The messageId to migrate from
     * @param taskId - The taskId to migrate to
     */
    migrateHttpRequestContextFromMessageToTask(messageId, taskId) {
        const ctx = this.getHttpRequestContextForMessage(messageId);
        if (ctx) {
            this.setHttpRequestContextForTask(taskId, ctx);
            this.deleteHttpRequestContextForMessage(messageId);
        }
    }
    /**
     * Deletes the HTTP context associated with a messageId.
     * @param messageId - The messageId to delete context for
     */
    deleteHttpRequestContextForMessage(messageId) {
        this.httpContextByMessageId.delete(messageId);
    }
    /**
     * Protected getter to access the private taskStore property from the parent class.
     * This is a workaround due to SDK limitations.
     */
    getTaskStore() {
        return this.taskStore;
    }
    /**
     * Protected getter to access the private eventBusManager property from the parent class.
     * This is a workaround due to SDK limitations.
     */
    getEventBusManager() {
        return this.eventBusManager;
    }
    /**
     * Protected getter to access the private _createRequestContext method from the parent class.
     * This is a workaround due to SDK limitations.
     */
    async callCreateRequestContext(incomingMessage, taskId, isStream) {
        return await this._createRequestContext(incomingMessage, taskId, isStream);
    }
    /**
     * Protected getter to access the private _processEvents method from the parent class.
     * This is a workaround due to SDK limitations.
     */
    callProcessEvents(...args) {
        return this._processEvents.apply(this, args);
    }
    /**
     * Resubscribes to a task's event stream, ensuring the task has updated metadata before yielding.
     * This method overrides the parent implementation to ensure metadata is updated before yielding.
     * @param params - Parameters containing the taskId
     * @returns Async generator of events
     */
    async *resubscribe(params) {
        const task = await this.getTaskStore().load(params.id);
        if (!task) {
            throw A2AError.taskNotFound(params.id);
        }
        // Yield task immediately (with current metadata)
        yield task;
        const finalStates = ['completed', 'failed', 'canceled', 'rejected'];
        if (finalStates.includes(task.status.state)) {
            return;
        }
        const eventBus = this.getEventBusManager().getByTaskId(params.id);
        if (!eventBus) {
            console.warn(`Resubscribe: No active event bus for task ${params.id}.`);
            return;
        }
        const eventQueue = new ExecutionEventQueue(eventBus);
        try {
            for await (const event of eventQueue.events()) {
                // Process event with ResultManager to ensure task is saved
                const resultManager = new ResultManager(this.getTaskStore());
                // Set context from task history if available
                if (task.history && task.history.length > 0) {
                    resultManager.setContext(task.history[0]);
                }
                await resultManager.processEvent(event);
                // Handle redemption and push notification for final status-update events
                if (event.kind === 'status-update' &&
                    event.final &&
                    event.taskId === params.id &&
                    terminalStates.includes(event.status?.state)) {
                    // Get HTTP context for this task
                    const httpContext = this.getHttpRequestContextForTask(event.taskId);
                    if (httpContext) {
                        const { bearerToken, validation } = httpContext;
                        if (bearerToken && validation) {
                            // Handle task finalization (redemption and push notification)
                            await this.handleTaskFinalization(resultManager, event, bearerToken, httpContext);
                        }
                    }
                }
                // Yield event after processing (so metadata is updated)
                if (event.kind === 'status-update' && event.taskId === params.id) {
                    yield event;
                }
                else if (event.kind === 'artifact-update' && event.taskId === params.id) {
                    yield event;
                }
                else if (event.kind === 'task' && event.id === params.id) {
                    yield event;
                }
            }
        }
        finally {
            eventQueue.stop();
        }
    }
}
