// @nevermined-io/payments/dist/x402/facilitator-api.js
/**
 * The FacilitatorAPI class provides methods to verify and settle AI agent permissions using X402 access tokens.
 * This allows AI agents to act as facilitators, verifying and settling credits on behalf of subscribers.
 *
 * @example
 * ```typescript
 * import { Payments, X402PaymentRequired } from '@nevermined-io/payments'
 *
 * // Initialize the Payments instance
 * const payments = Payments.getInstance({
 *   nvmApiKey: 'your-nvm-api-key',
 *   environment: 'sandbox'
 * })
 *
 * // The server's 402 PaymentRequired response
 * const paymentRequired: X402PaymentRequired = buildPaymentRequired('123456789', {
 *   endpoint: '/api/v1/agents/task',
 *   agentId: '987654321',
 *   httpVerb: 'POST'
 * })
 *
 * // Get X402 access token from subscriber (x402 v2: payment-signature header)
 * const x402Token = req.headers['payment-signature'] as string
 *
 * // Verify if subscriber has sufficient permissions/credits
 * const verification = await payments.facilitator.verifyPermissions({
 *   paymentRequired,
 *   x402AccessToken: x402Token,
 *   maxAmount: 2n
 * })
 *
 * if (verification.isValid) {
 *   // Settle (burn) the credits
 *   const settlement = await payments.facilitator.settlePermissions({
 *     paymentRequired,
 *     x402AccessToken: x402Token,
 *     maxAmount: 2n
 *   })
 *   console.log(`Credits redeemed: ${settlement.creditsRedeemed}`)
 * }
 * ```
 */
import { BasePaymentsAPI } from '../api/base-payments.js';
import { API_URL_SETTLE_PERMISSIONS, API_URL_VERIFY_PERMISSIONS } from '../api/nvm-api.js';
import { PaymentsError } from '../common/payments.error.js';
import { getDefaultNetwork } from '../common/types.js';
/**
 * Build an X402PaymentRequired object for verify/settle operations.
 *
 * This helper simplifies the creation of payment requirement objects
 * that are needed for the facilitator API.
 *
 * @param planId - The Nevermined plan identifier (required)
 * @param options - Optional configuration with endpoint, agentId, httpVerb, network, description
 * @returns X402PaymentRequired object ready to use with verifyPermissions/settlePermissions
 *
 * @example
 * ```typescript
 * import { buildPaymentRequired } from '@nevermined-io/payments'
 *
 * const paymentRequired = buildPaymentRequired('123456789', {
 *   endpoint: '/api/v1/agents/task',
 *   agentId: '987654321',
 *   httpVerb: 'POST'
 * })
 *
 * const result = await payments.facilitator.verifyPermissions({
 *   paymentRequired,
 *   x402AccessToken: token,
 *   maxAmount: 2n
 * })
 * ```
 */
export function buildPaymentRequired(planId, options) {
    const { endpoint, agentId, httpVerb, scheme = 'nvm:erc4337', network, description, environment, } = options || {};
    const resolvedNetwork = network ?? getDefaultNetwork(scheme, environment);
    // Build extra fields if any are provided
    const extra = agentId || httpVerb
        ? {
            ...(agentId && { agentId }),
            ...(httpVerb && { httpVerb }),
        }
        : undefined;
    return {
        x402Version: 2,
        resource: {
            url: endpoint || '',
            ...(description && { description }),
        },
        accepts: [
            {
                scheme,
                network: resolvedNetwork,
                planId,
                ...(extra && { extra }),
            },
        ],
        extensions: {},
    };
}
const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes
const planMetadataCache = new Map();
async function fetchPlanMetadata(payments, planId) {
    const cached = planMetadataCache.get(planId);
    if (cached && Date.now() - cached.cachedAt < CACHE_TTL_MS) {
        return { scheme: cached.scheme };
    }
    try {
        const plan = await payments.plans.getPlan(planId);
        const isCrypto = plan.registry?.price?.isCrypto;
        const scheme = isCrypto === false ? 'nvm:card-delegation' : 'nvm:erc4337';
        planMetadataCache.set(planId, { scheme, cachedAt: Date.now() });
        return { scheme };
    }
    catch {
        return { scheme: 'nvm:erc4337' };
    }
}
/**
 * Resolve the x402 scheme for a plan by fetching plan metadata (cached).
 * Used in callsites that don't have a token to extract scheme from
 * (402 responses and token generation).
 *
 * @param payments - The Payments instance for API access
 * @param planId - The plan identifier
 * @param explicitScheme - Optional explicit override; returned immediately if provided
 * @returns The resolved scheme type
 */
export async function resolveScheme(payments, planId, explicitScheme) {
    if (explicitScheme)
        return explicitScheme;
    const metadata = await fetchPlanMetadata(payments, planId);
    return metadata.scheme;
}
/**
 * The FacilitatorAPI class provides methods to verify and settle AI agent permissions.
 * It enables AI agents to act as facilitators, managing credit verification and settlement
 * for subscribers using X402 access tokens.
 */
export class FacilitatorAPI extends BasePaymentsAPI {
    /**
     * Get a singleton instance of the FacilitatorAPI class.
     *
     * @param options - The options to initialize the payments class
     * @returns The instance of the FacilitatorAPI class
     */
    static getInstance(options) {
        return new FacilitatorAPI(options);
    }
    /**
     * Verify if a subscriber has permission to use credits from a payment plan.
     * This method simulates the credit usage without actually burning credits,
     * checking if the subscriber has sufficient balance and permissions.
     *
     * The planId and subscriberAddress are extracted from the x402AccessToken.
     *
     * @param params - Verification parameters (see {@link VerifyPermissionsParams}).
     *   - paymentRequired: x402 PaymentRequired from 402 response (required, for validation)
     *   - x402AccessToken: X402 access token (contains planId, subscriberAddress, agentId)
     *   - maxAmount: maximum credits to verify (optional, bigint)
     * @returns A promise that resolves to a verification result with 'isValid' boolean
     *
     * @throws PaymentsError if verification fails
     */
    async verifyPermissions(params) {
        const { paymentRequired, x402AccessToken, maxAmount } = params;
        const url = new URL(API_URL_VERIFY_PERMISSIONS, this.environment.backend);
        const body = {
            paymentRequired,
            x402AccessToken,
        };
        if (maxAmount !== undefined) {
            body.maxAmount = maxAmount.toString();
        }
        const options = this.getPublicHTTPOptions('POST', body);
        try {
            const response = await fetch(url, options);
            if (!response.ok) {
                let errorMessage = 'Permission verification failed';
                try {
                    const errorData = await response.json();
                    errorMessage = errorData.message || errorMessage;
                }
                catch {
                    // Use default error message
                }
                throw PaymentsError.fromBackend(errorMessage, {
                    message: errorMessage,
                    code: `HTTP ${response.status}`,
                });
            }
            return await response.json();
        }
        catch (error) {
            if (error instanceof PaymentsError) {
                throw error;
            }
            throw PaymentsError.fromBackend('Network error during permission verification', {
                message: error instanceof Error ? error.message : String(error),
                code: 'network_error',
            });
        }
    }
    /**
     * Settle (burn) credits from a subscriber's payment plan.
     * This method executes the actual credit consumption, burning the specified
     * number of credits from the subscriber's balance. If the subscriber doesn't
     * have enough credits, it will attempt to order more before settling.
     *
     * The planId and subscriberAddress are extracted from the x402AccessToken.
     *
     * @param params - Settlement parameters (see {@link SettlePermissionsParams}).
     *   - paymentRequired: x402 PaymentRequired from 402 response (required, for validation)
     *   - x402AccessToken: X402 access token (contains planId, subscriberAddress, agentId)
     *   - maxAmount: number of credits to burn (optional, bigint)
     *   - agentRequestId: Agent request ID for observability tracking (optional)
     *   - batch: Whether this is a batch request (optional)
     *   - marginPercent: Margin percentage for credit calculation (optional)
     * @returns A promise that resolves to a settlement result with transaction details
     *
     * @throws PaymentsError if settlement fails
     */
    async settlePermissions(params) {
        const { paymentRequired, x402AccessToken, maxAmount, agentRequestId, batch, marginPercent } = params;
        const url = new URL(API_URL_SETTLE_PERMISSIONS, this.environment.backend);
        const body = {
            paymentRequired,
            x402AccessToken,
        };
        if (maxAmount !== undefined) {
            body.maxAmount = maxAmount.toString();
        }
        if (agentRequestId !== undefined) {
            body.agentRequestId = agentRequestId;
        }
        if (batch !== undefined) {
            body.batch = batch;
        }
        if (marginPercent !== undefined) {
            body.marginPercent = marginPercent;
        }
        const options = this.getPublicHTTPOptions('POST', body);
        try {
            const response = await fetch(url, options);
            if (!response.ok) {
                let errorMessage = 'Permission settlement failed';
                try {
                    const errorData = await response.json();
                    errorMessage = errorData.message || errorMessage;
                }
                catch {
                    // Use default error message
                }
                throw PaymentsError.fromBackend(errorMessage, {
                    message: errorMessage,
                    code: `HTTP ${response.status}`,
                });
            }
            return await response.json();
        }
        catch (error) {
            if (error instanceof PaymentsError) {
                throw error;
            }
            throw PaymentsError.fromBackend('Network error during permission settlement', {
                message: error instanceof Error ? error.message : String(error),
                code: 'network_error',
            });
        }
    }
}
