// @nevermined-io/payments/dist/mcp/core/paywall.js
/**
 * Main paywall decorator for MCP handlers (tools, resources, prompts)
 */
import { isValidScheme } from '../../common/types.js';
import { decodeAccessToken } from '../../utils.js';
import { buildPaymentRequired, } from '../../x402/facilitator-api.js';
import { ERROR_CODES, createRpcError } from '../utils/errors.js';
/**
 * Main class for creating paywall-protected MCP handlers
 */
export class PaywallDecorator {
    constructor(payments, authenticator, creditsContext) {
        this.payments = payments;
        this.authenticator = authenticator;
        this.creditsContext = creditsContext;
        // Internal config ensures serverName is always a concrete string
        this.config = {
            agentId: '',
            serverName: 'mcp-server',
        };
    }
    /**
     * Configure the paywall with agent and server information
     */
    configure(options) {
        this.config = {
            agentId: options.agentId || this.config.agentId,
            serverName: options.serverName ?? this.config.serverName,
        };
    }
    protect(handler, options) {
        return this.createWrappedHandler(handler, options);
    }
    /**
     * Internal method to create the wrapped handler
     */
    createWrappedHandler(handler, options) {
        return async (...allArgs) => {
            // Validate configuration
            if (!this.config.agentId) {
                throw createRpcError(ERROR_CODES.Misconfiguration, 'Server misconfiguration: missing agentId');
            }
            const kind = options?.kind ?? 'tool';
            const name = options?.name ?? 'unnamed';
            // Detect resource signature: (url, variables, extra)
            const isResource = allArgs.length >= 2 && allArgs[0] instanceof URL;
            const extra = isResource ? allArgs[2] : allArgs[1];
            const argsOrVars = isResource ? allArgs[1] : allArgs[0];
            // 1. Authenticate request
            const authResult = await this.authenticator.authenticate(extra, { planId: options?.planId, maxAmount: options?.maxAmount }, this.config.agentId, this.config.serverName, name, kind, argsOrVars);
            // 2. Pre-calculate credits if they are fixed (not a function)
            // This allows handlers to access credits during execution
            const creditsOption = options?.credits;
            const isFixedCredits = typeof creditsOption === 'bigint' || creditsOption === undefined;
            const preCalculatedCredits = isFixedCredits
                ? this.creditsContext.resolve(creditsOption, argsOrVars, null, authResult)
                : undefined;
            // Determine effective planId: explicit option overrides token-derived value
            const effectivePlanId = options?.planId ?? authResult.planId;
            // 3. Build PaywallContext for handler (with extra wrapper for backward compatibility)
            const paywallContext = {
                authResult,
                credits: preCalculatedCredits,
                planId: authResult.planId,
                subscriberAddress: authResult.subscriberAddress,
                agentRequest: authResult.agentRequest,
            };
            // 4. Execute original handler with context
            const result = await handler(...allArgs, paywallContext);
            // 5. Resolve final credits to burn (may be different if credits are dynamic)
            const credits = isFixedCredits
                ? (preCalculatedCredits ?? 1n)
                : this.creditsContext.resolve(creditsOption, argsOrVars, result, authResult);
            // Update context with final resolved credits
            paywallContext.credits = credits;
            // 6. If the result is an AsyncIterable (stream), redeem on completion
            if (isAsyncIterable(result)) {
                const onFinally = async () => {
                    return await this.redeemCredits(effectivePlanId, authResult.token, authResult.subscriberAddress, credits, options, authResult.agentId, authResult.logicalUrl, authResult.httpUrl, 'POST');
                };
                return wrapAsyncIterable(result, onFinally, effectivePlanId, authResult.subscriberAddress, credits);
            }
            // 7. Non-streaming: redeem immediately
            const creditsResult = await this.redeemCredits(effectivePlanId, authResult.token, authResult.subscriberAddress, credits, options, authResult.agentId, authResult.logicalUrl, 'POST', authResult.httpUrl);
            result._meta = {
                ...result._meta,
                ...(creditsResult.transaction && { txHash: creditsResult.transaction }),
                creditsRedeemed: creditsResult.success ? (creditsResult.creditsRedeemed ?? credits.toString()) : '0',
                remainingBalance: creditsResult.remainingBalance,
                planId: authResult.planId,
                subscriberAddress: authResult.subscriberAddress,
                success: creditsResult.success,
                ...(creditsResult.errorReason && { errorReason: creditsResult.errorReason }),
            };
            return result;
        };
    }
    /**
     * Redeem credits after successful request
     */
    async redeemCredits(planId, token, subscriberAddress, credits, options, agentId, endpoint, fallbackEndpoint, httpVerb) {
        let ret = {
            success: true,
            transaction: '',
            network: '',
        };
        const decoded = decodeAccessToken(token);
        const scheme = isValidScheme(decoded?.accepted?.scheme) ? decoded.accepted.scheme : 'nvm:erc4337';
        try {
            if (credits && credits > 0n && subscriberAddress && planId) {
                const paymentRequired = buildPaymentRequired(planId, {
                    endpoint: endpoint || '',
                    agentId,
                    httpVerb,
                    scheme,
                    environment: this.payments.getEnvironmentName(),
                });
                ret = await this.payments.facilitator.settlePermissions({
                    paymentRequired,
                    x402AccessToken: token,
                    maxAmount: credits,
                });
            }
        }
        catch (primaryError) {
            // If logical URL fails and we have an HTTP URL fallback, retry with it
            let lastError = primaryError;
            if (fallbackEndpoint) {
                try {
                    const paymentRequired = buildPaymentRequired(planId, {
                        endpoint: fallbackEndpoint,
                        agentId,
                        httpVerb,
                        scheme,
                        environment: this.payments.getEnvironmentName(),
                    });
                    ret = await this.payments.facilitator.settlePermissions({
                        paymentRequired,
                        x402AccessToken: token,
                        maxAmount: credits,
                    });
                    return ret;
                }
                catch (fallbackError) {
                    // Fallback also failed, use fallback error as the reported error
                    lastError = fallbackError;
                }
            }
            ret.success = false;
            ret.errorReason = lastError instanceof Error ? lastError.message : String(lastError);
            if (options.onRedeemError === 'propagate') {
                throw createRpcError(ERROR_CODES.Misconfiguration, `Failed to redeem credits: ${ret.errorReason}`);
            }
            // Default: attach error to result but don't throw
        }
        return ret;
    }
}
/**
 * Type guard to detect AsyncIterable values.
 */
function isAsyncIterable(value) {
    return value != null && typeof value[Symbol.asyncIterator] === 'function';
}
/**
 * Wrap an AsyncIterable with metadata injection at the end of the stream
 */
function wrapAsyncIterable(iterable, onFinally, planId, subscriberAddress, credits) {
    async function* generator() {
        let creditsResult = null;
        try {
            for await (const chunk of iterable) {
                yield chunk;
            }
        }
        finally {
            creditsResult = await onFinally();
        }
        // Yield a _meta chunk at the end with the redemption result
        const metadataChunk = {
            _meta: {
                // Only include txHash if it has a value
                ...(creditsResult?.transaction && { txHash: creditsResult.transaction }),
                creditsRedeemed: creditsResult?.success ? (creditsResult.creditsRedeemed ?? credits.toString()) : '0',
                remainingBalance: creditsResult?.remainingBalance,
                planId,
                subscriberAddress,
                success: creditsResult?.success || false,
                ...(creditsResult?.errorReason && { errorReason: creditsResult.errorReason }),
            },
        };
        yield metadataChunk;
    }
    return generator();
}
