// @nevermined-io/payments/dist/common/types.js
/**
 * Different types of prices that can be configured for a plan
 * @remarks 0 - FIXED_PRICE, 1 - FIXED_FIAT_PRICE, 2 - SMART_CONTRACT_PRICE
 * If FIXED_PRICE it means the plan can be paid in crypto by a fixed amount of a ERC20 or Native token
 * If FIXED_FIAT_PRICE it means the plan can be paid in fiat by a fixed amount (typically USD)
 * If SMART_CONTRACT_PRICE it means the plan can be paid in crypto and the amount to be paid is calculated by a smart contract
 */
export var PlanPriceType;
(function (PlanPriceType) {
    PlanPriceType[PlanPriceType["FIXED_PRICE"] = 0] = "FIXED_PRICE";
    PlanPriceType[PlanPriceType["FIXED_FIAT_PRICE"] = 1] = "FIXED_FIAT_PRICE";
    PlanPriceType[PlanPriceType["SMART_CONTRACT_PRICE"] = 2] = "SMART_CONTRACT_PRICE";
})(PlanPriceType || (PlanPriceType = {}));
/**
 * Different types of credits that can be obtained when purchasing a plan
 * @remarks 0 - EXPIRABLE, 1 - FIXED, 2 - DYNAMIC
 * If EXPIRABLE it means the credits can be used for a fixed amount of time (calculated in seconds)
 * If FIXED it means the credits can be used for a fixed amount of times
 * If DYNAMIC it means the credits can be used but the redemption amount is dynamic
 */
export var PlanCreditsType;
(function (PlanCreditsType) {
    PlanCreditsType[PlanCreditsType["EXPIRABLE"] = 0] = "EXPIRABLE";
    PlanCreditsType[PlanCreditsType["FIXED"] = 1] = "FIXED";
    PlanCreditsType[PlanCreditsType["DYNAMIC"] = 2] = "DYNAMIC";
})(PlanCreditsType || (PlanCreditsType = {}));
/**
 * Different types of redemptions criterias that can be used when redeeming credits
 * @remarks 0 - ONLY_GLOBAL_ROLE, 1 - ONLY_OWNER, 2 - ROLE_AND_OWNER
 * If ONLY_GLOBAL_ROLE it means the credits can be redeemed only by an account with the `CREDITS_BURNER_ROLE`
 * If ONLY_OWNER it means the credits can be redeemed only by the owner of the Plan
 * If ONLY_PLAN_ROLE it means the credits can be redeemed by an account with specifics grants for the plan
 * If ONLY_SUBSCRIBER means only the subscriber (the one who holds the credits NFT) can redeem
 */
export var PlanRedemptionType;
(function (PlanRedemptionType) {
    PlanRedemptionType[PlanRedemptionType["ONLY_GLOBAL_ROLE"] = 0] = "ONLY_GLOBAL_ROLE";
    PlanRedemptionType[PlanRedemptionType["ONLY_OWNER"] = 1] = "ONLY_OWNER";
    PlanRedemptionType[PlanRedemptionType["ONLY_PLAN_ROLE"] = 2] = "ONLY_PLAN_ROLE";
    PlanRedemptionType[PlanRedemptionType["ONLY_SUBSCRIBER"] = 4] = "ONLY_SUBSCRIBER";
})(PlanRedemptionType || (PlanRedemptionType = {}));
/**
 * Options for pagination in API requests to the Nevermined API.
 */
export class PaginationOptions {
    /**
     * Constructs a new PaginationOptions instance.
     * @param options - Optional initial values for the pagination options.
     */
    constructor(options) {
        /**
         * The order in which to sort the results.
         * Default is 'desc' (descending).
         */
        this.sortOrder = 'desc';
        /**
         * The page number to retrieve.
         * Default is 1.
         */
        this.page = 1;
        /**
         * The number of items per page.
         * Default is 10.
         */
        this.offset = 10;
        if (options) {
            this.sortBy = options.sortBy;
            this.sortOrder = options.sortOrder || 'desc';
            this.page = options.page || 1;
            this.offset = options.offset || 10;
        }
    }
    /**
     * It returns a string representation of the pagination options
     * @returns A string representation of the pagination options as URL query parameters.
     * This can be used to append to API requests for pagination.
     */
    asQueryParams() {
        const params = {};
        if (this.sortBy) {
            params.sortBy = this.sortBy;
        }
        params.sortOrder = this.sortOrder;
        params.page = this.page.toString();
        params.pageSize = this.offset.toString();
        return new URLSearchParams(params).toString();
    }
}
/**
 * Status of an agent task
 */
export var AgentTaskStatus;
(function (AgentTaskStatus) {
    AgentTaskStatus["SUCCESS"] = "SUCCESS";
    AgentTaskStatus["FAILED"] = "FAILED";
    AgentTaskStatus["PENDING"] = "PENDING";
})(AgentTaskStatus || (AgentTaskStatus = {}));
/**
 * Default network for each x402 scheme (backward-compat: defaults to Base Sepolia).
 */
export const X402_SCHEME_NETWORKS = {
    'nvm:erc4337': 'eip155:84532',
    'nvm:card-delegation': 'stripe',
};
/**
 * Environment-specific network for the erc4337 scheme.
 */
const ERC4337_NETWORK_BY_ENV = {
    sandbox: 'eip155:84532', // Base Sepolia
    staging_sandbox: 'eip155:84532', // Base Sepolia
    live: 'eip155:8453', // Base Mainnet
    staging_live: 'eip155:8453', // Base Mainnet
};
/**
 * Return the default network for a scheme, accounting for the environment.
 *
 * For `nvm:erc4337`, the network depends on the environment:
 * - sandbox / staging_sandbox → `eip155:84532` (Base Sepolia)
 * - live / staging_live → `eip155:8453` (Base Mainnet)
 *
 * Falls back to `X402_SCHEME_NETWORKS` when no environment is given.
 */
export function getDefaultNetwork(scheme, environment) {
    if (scheme === 'nvm:erc4337' && environment) {
        return ERC4337_NETWORK_BY_ENV[environment] ?? 'eip155:84532';
    }
    return X402_SCHEME_NETWORKS[scheme];
}
/**
 * Type guard to check if a value is a valid x402 scheme type.
 */
export function isValidScheme(s) {
    return s === 'nvm:erc4337' || s === 'nvm:card-delegation';
}
