// @traceloop/node-server-sdk/dist/index.mjs
import { NodeSDK } from '@opentelemetry/sdk-node';
import { diag, DiagConsoleLogger, DiagLogLevel, createContextKey, trace, context } from '@opentelemetry/api';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-proto';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { ATTR_SERVICE_NAME } from '@opentelemetry/semantic-conventions';
import * as os from 'os';
import * as fs from 'fs';
import * as path from 'path';
import { v4 } from 'uuid';
import { PostHog } from 'posthog-node';
import fetch$1 from 'cross-fetch';
import fetchBuilder from 'fetch-retry';
import { SpanAttributes, CONTEXT_KEY_ALLOW_TRACE_CONTENT, TraceloopSpanKindValues, Events, EventAttributes } from '@traceloop/ai-semantic-conventions';
import { AnthropicInstrumentation } from '@traceloop/instrumentation-anthropic';
import { OpenAIInstrumentation } from '@traceloop/instrumentation-openai';
import { LlamaIndexInstrumentation } from '@traceloop/instrumentation-llamaindex';
import { VertexAIInstrumentation, AIPlatformInstrumentation } from '@traceloop/instrumentation-vertexai';
import { BedrockInstrumentation } from '@traceloop/instrumentation-bedrock';
import { CohereInstrumentation } from '@traceloop/instrumentation-cohere';
import { PineconeInstrumentation } from '@traceloop/instrumentation-pinecone';
import { LangChainInstrumentation } from '@traceloop/instrumentation-langchain';
import { ChromaDBInstrumentation } from '@traceloop/instrumentation-chromadb';
import { QdrantInstrumentation } from '@traceloop/instrumentation-qdrant';
import { TogetherInstrumentation } from '@traceloop/instrumentation-together';
import { SimpleSpanProcessor, BatchSpanProcessor } from '@opentelemetry/sdk-trace-node';
import { suppressTracing } from '@opentelemetry/core';
import { Environment } from 'nunjucks';

var version = "0.14.6";

class Telemetry {
    static getInstance() {
        if (!Telemetry.instance) {
            Telemetry.instance = new Telemetry();
        }
        return Telemetry.instance;
    }
    constructor() {
        this.telemetryEnabled =
            !process.env.TRACELOOP_TELEMETRY ||
                process.env.TRACELOOP_TELEMETRY.toLowerCase() === "true";
        if (this.telemetryEnabled) {
            this.posthog = new PostHog("phc_JMTeAfG8OpaPsyHzSBtqquMvko1fmOHcW0gyqLCrF3t");
        }
    }
    getAnonId() {
        if (this.anonId) {
            return this.anonId;
        }
        try {
            if (!fs.existsSync(Telemetry.ANON_ID_PATH)) {
                fs.mkdirSync(path.dirname(Telemetry.ANON_ID_PATH), { recursive: true });
                const anonIdFile = fs.openSync(Telemetry.ANON_ID_PATH, "w");
                this.anonId = v4();
                fs.writeSync(anonIdFile, this.anonId);
                fs.closeSync(anonIdFile);
            }
            else {
                const anonIdFile = fs.openSync(Telemetry.ANON_ID_PATH, "r");
                this.anonId = fs.readFileSync(anonIdFile, "utf8");
                fs.closeSync(anonIdFile);
            }
            return this.anonId;
        }
        catch (e) {
            return Telemetry.UNKNOWN_ANON_ID;
        }
    }
    getContext() {
        return {
            sdk: "typescript",
            sdk_version: version,
        };
    }
    capture(event, properties) {
        if (this.telemetryEnabled && this.posthog) {
            this.posthog.capture({
                distinctId: this.getAnonId(),
                event,
                properties: Object.assign(Object.assign({}, properties), this.getContext()),
            });
            this.posthog.flush();
        }
    }
    logException(error) {
        if (this.telemetryEnabled) {
            this.capture("error", { error: error.message, stack: error.stack || "" });
        }
    }
}
Telemetry.ANON_ID_PATH = `${os.homedir()}/.cache/traceloop/telemetry_anon_id`;
Telemetry.UNKNOWN_ANON_ID = "UNKNOWN";

/**
 * The severity of an error.
 */
const SEVERITY = {
    Warning: "Warning",
    Error: "Error",
    Critical: "Critical",
};
/**
 * Base class for all Traceloop errors.
 */
class TraceloopError extends Error {
    constructor(message, severity = SEVERITY.Error) {
        super(message);
        this.severity = severity;
    }
}
class NotInitializedError extends TraceloopError {
    constructor() {
        super(`The Traceloop SDK must be initialized by calling the "initialize" function prior to use.`, SEVERITY.Critical);
    }
}
class InitializationError extends TraceloopError {
    constructor(message, cause) {
        super(message !== null && message !== void 0 ? message : "Failed to initialize Traceloop SDK", SEVERITY.Critical);
        this.underlyingCause = cause;
    }
}
class ArgumentNotProvidedError extends TraceloopError {
    constructor(argumentName) {
        super(`The "${argumentName}" argument is required and must be a string.`);
    }
}
class PromptNotFoundError extends TraceloopError {
    constructor(key) {
        super(`The prompt "${key}" was not found in the registry.`);
    }
}

const validateConfiguration = (options) => {
    const { apiKey, traceloopSyncEnabled, traceloopSyncMaxRetries, traceloopSyncPollingInterval, traceloopSyncDevPollingInterval, } = options;
    if (apiKey && typeof apiKey !== "string") {
        throw new InitializationError('"apiKey" must be a string');
    }
    if (traceloopSyncEnabled) {
        if (typeof traceloopSyncMaxRetries !== "number" ||
            traceloopSyncMaxRetries <= 0) {
            throw new InitializationError('"traceloopSyncMaxRetries" must be an integer greater than 0.');
        }
        if (typeof traceloopSyncPollingInterval !== "number" ||
            traceloopSyncPollingInterval <= 0) {
            throw new InitializationError('"traceloopSyncPollingInterval" must be an integer greater than 0.');
        }
        if (typeof traceloopSyncDevPollingInterval !== "number" ||
            traceloopSyncDevPollingInterval <= 0) {
            throw new InitializationError('"traceloopSyncDevPollingInterval" must be an integer greater than 0.');
        }
    }
};

const fetchRetry = fetchBuilder(fetch$1);
const fetchPrompts = async (options) => {
    const { apiKey, baseUrl, traceloopSyncMaxRetries } = options;
    const response = await fetchRetry(`${baseUrl}/v1/traceloop/prompts`, {
        method: "GET",
        headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${apiKey}`,
            "X-Traceloop-SDK-Version": "0.0.30",
        },
        retries: traceloopSyncMaxRetries,
        retryOn: function (attempt, error, response) {
            if (attempt >= traceloopSyncMaxRetries)
                return false;
            if ((response === null || response === void 0 ? void 0 : response.status) && response.status >= 500) {
                return true;
            }
            return false;
        },
        retryDelay: function (attempt) {
            return Math.pow(2, attempt) * 1000; // 1000, 2000, 4000
        },
    });
    return await response.json();
};

const _prompts = {};
let _initialized = false;
let _initializedPromise;
/**
 * Returns true once SDK prompt registry has been initialized, else rejects with an error.
 * @returns Promise<boolean>
 */
const waitForInitialization = async () => {
    if (_initialized) {
        return true;
    }
    return await _initializedPromise;
};
const getPromptByKey = (key) => {
    if (!_prompts[key]) {
        throw new PromptNotFoundError(key);
    }
    return _prompts[key];
};
const populateRegistry = (prompts) => {
    prompts === null || prompts === void 0 ? void 0 : prompts.forEach((prompt) => {
        _prompts[prompt.key] = prompt;
    });
};
const initializeRegistry = (options) => {
    const { baseUrl, traceloopSyncEnabled, traceloopSyncPollingInterval, traceloopSyncDevPollingInterval, } = options;
    if (!traceloopSyncEnabled || !(baseUrl === null || baseUrl === void 0 ? void 0 : baseUrl.includes("traceloop")))
        return;
    let pollingInterval = traceloopSyncPollingInterval;
    _initializedPromise = fetchPrompts(options)
        .then(({ prompts, environment }) => {
        if (environment === "dev") {
            pollingInterval = traceloopSyncDevPollingInterval;
        }
        populateRegistry(prompts);
        _initialized = true;
        setInterval(async () => {
            try {
                const { prompts } = await fetchPrompts(options);
                populateRegistry(prompts);
            }
            catch (err) {
                diag.error("Failed to fetch prompt data", err);
            }
        }, pollingInterval * 1000).unref();
        return true;
    })
        .catch((e) => {
        throw new InitializationError("Failed to fetch prompt data to initialize Traceloop SDK", e);
    });
};

/**
 * Base class for handling annotation operations with the Traceloop API.
 * @internal
 */
class BaseAnnotation {
    constructor(client, flow) {
        this.client = client;
        this.flow = flow;
    }
    /**
     * Creates a new annotation.
     *
     * @param options - The annotation creation options
     * @returns Promise resolving to the fetch Response
     */
    async create(options) {
        return await this.client.post(`/v2/annotation-tasks/${options.annotationTask}/annotations`, {
            entity_instance_id: options.entity.id,
            tags: options.tags,
            source: "sdk",
            flow: this.flow,
            actor: {
                type: "service",
                id: this.client.appName,
            },
        });
    }
}

/**
 * Handles user feedback annotations with the Traceloop API.
 */
class UserFeedback extends BaseAnnotation {
    constructor(client) {
        super(client, "user_feedback");
    }
    /**
     * Creates a new annotation for a specific task and entity.
     *
     * @param options - The options for creating an annotation
     * @returns Promise resolving to the fetch Response
     *
     * @example
     * ```typescript
     * await client.annotation.create({
     *   annotationTask: 'sample-annotation-task',
     *   entity: {
     *     id: '123456',
     *   },
     *   tags: {
     *     sentiment: 'positive',
     *     score: 0.85,
     *     tones: ['happy', 'surprised']
     *   }
     * });
     * ```
     */
    async create(options) {
        return await super.create(options);
    }
}

/**
 * The main client for interacting with Traceloop's API.
 * This client can be used either directly or through the singleton pattern via configuration.
 *
 * @example
 * // Direct usage
 * const client = new TraceloopClient('your-api-key');
 *
 * @example
 * // Through configuration (recommended)
 * initialize({ apiKey: 'your-api-key', appName: 'your-app' });
 * const client = getClient();
 */
class TraceloopClient {
    /**
     * Creates a new instance of the TraceloopClient.
     *
     * @param options - Configuration options for the client
     */
    constructor(options) {
        this.version = version;
        this.userFeedback = new UserFeedback(this);
        this.apiKey = options.apiKey;
        this.appName = options.appName;
        this.baseUrl =
            options.baseUrl ||
                process.env.TRACELOOP_BASE_URL ||
                "https://api.traceloop.com";
    }
    async post(path, body) {
        return await fetch(`${this.baseUrl}${path}`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${this.apiKey}`,
                "X-Traceloop-SDK-Version": this.version,
            },
            body: JSON.stringify(body),
        });
    }
}

let _configuration;
let _client;
/**
 * Initializes the Traceloop SDK and creates a singleton client instance if API key is provided.
 * Must be called once before any other SDK methods.
 *
 * @param options - The options to initialize the SDK. See the {@link InitializeOptions} for details.
 * @returns TraceloopClient - The singleton client instance if API key is provided, otherwise undefined.
 * @throws {InitializationError} if the configuration is invalid or if failed to fetch feature data.
 *
 * @example
 * ```typescript
 * initialize({
 *   apiKey: 'your-api-key',
 *   appName: 'your-app',
 * });
 * ```
 */
const initialize = (options = {}) => {
    if (_configuration) {
        return;
    }
    if (!options.baseUrl) {
        options.baseUrl =
            process.env.TRACELOOP_BASE_URL || "https://api.traceloop.com";
    }
    if (!options.apiKey) {
        options.apiKey = process.env.TRACELOOP_API_KEY;
    }
    if (!options.appName) {
        options.appName = process.env.npm_package_name;
    }
    if (options.traceloopSyncEnabled === undefined) {
        if (process.env.TRACELOOP_SYNC_ENABLED !== undefined) {
            options.traceloopSyncEnabled = ["1", "true"].includes(process.env.TRACELOOP_SYNC_ENABLED.toLowerCase());
        }
        else {
            options.traceloopSyncEnabled = true;
        }
    }
    if (options.traceloopSyncEnabled) {
        if (!options.traceloopSyncMaxRetries) {
            options.traceloopSyncMaxRetries =
                Number(process.env.TRACELOOP_SYNC_MAX_RETRIES) || 3;
        }
        if (!options.traceloopSyncPollingInterval) {
            options.traceloopSyncPollingInterval =
                Number(process.env.TRACELOOP_SYNC_POLLING_INTERVAL) || 60;
        }
        if (!options.traceloopSyncDevPollingInterval) {
            options.traceloopSyncDevPollingInterval =
                Number(process.env.TRACELOOP_SYNC_DEV_POLLING_INTERVAL) || 5;
        }
    }
    validateConfiguration(options);
    _configuration = Object.freeze(options);
    if (!options.silenceInitializationMessage) {
        console.log(`Traceloop exporting traces to ${_configuration.exporter ? "a custom exporter" : _configuration.baseUrl}`);
    }
    if (options.tracingEnabled === undefined || options.tracingEnabled) {
        if (options.logLevel) {
            diag.setLogger(new DiagConsoleLogger(), logLevelToOtelLogLevel(options.logLevel));
        }
        startTracing(_configuration);
    }
    initializeRegistry(_configuration);
    if (options.apiKey) {
        _client = new TraceloopClient({
            apiKey: options.apiKey,
            baseUrl: options.baseUrl,
            appName: options.appName,
        });
        return _client;
    }
    return;
};
const logLevelToOtelLogLevel = (logLevel) => {
    switch (logLevel) {
        case "debug":
            return DiagLogLevel.DEBUG;
        case "info":
            return DiagLogLevel.INFO;
        case "warn":
            return DiagLogLevel.WARN;
        case "error":
            return DiagLogLevel.ERROR;
    }
};
/**
 * Gets the singleton instance of the TraceloopClient.
 * The SDK must be initialized with an API key before calling this function.
 *
 * @returns The TraceloopClient singleton instance
 * @throws {Error} if the SDK hasn't been initialized or was initialized without an API key
 *
 * @example
 * ```typescript
 * const client = getClient();
 * await client.annotation.create({ annotationTask: 'taskId', entityInstanceId: 'entityId', tags: { score: 0.9 } });
 * ```
 */
const getClient = () => {
    if (!_client) {
        throw new Error("Traceloop must be initialized before getting client, Call initialize() first." +
            "If you already called initialize(), make sure you have an api key.");
    }
    return _client;
};

const TRACER_NAME = "@traceloop/node-server-sdk";
const WORKFLOW_NAME_KEY = createContextKey("workflow_name");
const ENTITY_NAME_KEY = createContextKey("entity_name");
const ASSOCATION_PROPERTIES_KEY = createContextKey("association_properties");
const getTracer = () => {
    return trace.getTracer(TRACER_NAME);
};
const getEntityPath = (entityContext) => {
    const path = entityContext.getValue(ENTITY_NAME_KEY);
    return path ? `${path}` : undefined;
};

const AI_GENERATE_TEXT_DO_GENERATE = "ai.generateText.doGenerate";
const AI_STREAM_TEXT_DO_STREAM = "ai.streamText.doStream";
const HANDLED_SPAN_NAMES = {
    [AI_GENERATE_TEXT_DO_GENERATE]: "ai.generateText.generate",
    [AI_STREAM_TEXT_DO_STREAM]: "ai.streamText.stream",
};
const AI_RESPONSE_TEXT = "ai.response.text";
const AI_PROMPT_MESSAGES = "ai.prompt.messages";
const AI_USAGE_PROMPT_TOKENS = "ai.usage.promptTokens";
const AI_USAGE_COMPLETION_TOKENS = "ai.usage.completionTokens";
const AI_MODEL_PROVIDER = "ai.model.provider";
const transformAiSdkSpanName = (span) => {
    // Unfortunately, the span name is not writable as this is not the intended behavior
    // but it is a workaround to set the correct span name
    if (span.name in HANDLED_SPAN_NAMES) {
        span.name = HANDLED_SPAN_NAMES[span.name];
    }
};
const transformResponseText = (attributes) => {
    if (AI_RESPONSE_TEXT in attributes) {
        attributes[`${SpanAttributes.LLM_COMPLETIONS}.0.content`] =
            attributes[AI_RESPONSE_TEXT];
        attributes[`${SpanAttributes.LLM_COMPLETIONS}.0.role`] = "assistant";
        delete attributes[AI_RESPONSE_TEXT];
    }
};
const transformPromptMessages = (attributes) => {
    if (AI_PROMPT_MESSAGES in attributes) {
        try {
            const messages = JSON.parse(attributes[AI_PROMPT_MESSAGES]);
            messages.forEach((msg, index) => {
                attributes[`${SpanAttributes.LLM_PROMPTS}.${index}.content`] =
                    typeof msg.content === "string"
                        ? msg.content
                        : JSON.stringify(msg.content);
                attributes[`${SpanAttributes.LLM_PROMPTS}.${index}.role`] = msg.role;
            });
            delete attributes[AI_PROMPT_MESSAGES];
        }
        catch (_a) {
            // Skip if JSON parsing fails
        }
    }
};
const transformPromptTokens = (attributes) => {
    if (AI_USAGE_PROMPT_TOKENS in attributes) {
        attributes[`${SpanAttributes.LLM_USAGE_PROMPT_TOKENS}`] =
            attributes[AI_USAGE_PROMPT_TOKENS];
        delete attributes[AI_USAGE_PROMPT_TOKENS];
    }
};
const transformCompletionTokens = (attributes) => {
    if (AI_USAGE_COMPLETION_TOKENS in attributes) {
        attributes[`${SpanAttributes.LLM_USAGE_COMPLETION_TOKENS}`] =
            attributes[AI_USAGE_COMPLETION_TOKENS];
        delete attributes[AI_USAGE_COMPLETION_TOKENS];
    }
};
const calculateTotalTokens = (attributes) => {
    const promptTokens = attributes[`${SpanAttributes.LLM_USAGE_PROMPT_TOKENS}`];
    const completionTokens = attributes[`${SpanAttributes.LLM_USAGE_COMPLETION_TOKENS}`];
    if (promptTokens && completionTokens) {
        attributes[`${SpanAttributes.LLM_USAGE_TOTAL_TOKENS}`] =
            Number(promptTokens) + Number(completionTokens);
    }
};
const transformVendor = (attributes) => {
    if (AI_MODEL_PROVIDER in attributes) {
        const vendor = attributes[AI_MODEL_PROVIDER];
        if (vendor && vendor.startsWith("openai")) {
            attributes[SpanAttributes.LLM_SYSTEM] = "OpenAI";
        }
        else {
            attributes[SpanAttributes.LLM_SYSTEM] = vendor;
        }
        delete attributes[AI_MODEL_PROVIDER];
    }
};
const transformAiSdkAttributes = (attributes) => {
    transformResponseText(attributes);
    transformPromptMessages(attributes);
    transformPromptTokens(attributes);
    transformCompletionTokens(attributes);
    calculateTotalTokens(attributes);
    transformVendor(attributes);
};
const shouldHandleSpan = (span) => {
    return span.name in HANDLED_SPAN_NAMES;
};
const transformAiSdkSpan = (span) => {
    if (!shouldHandleSpan(span)) {
        return;
    }
    transformAiSdkSpanName(span);
    transformAiSdkAttributes(span.attributes);
};

function parseKeyPairsIntoRecord(keyPairs) {
    const result = {};
    if (!keyPairs)
        return result;
    keyPairs.split(",").forEach((pair) => {
        const [key, value] = pair.split("=");
        if (key && value) {
            result[key.trim()] = value.trim();
        }
    });
    return result;
}

const ALL_INSTRUMENTATION_LIBRARIES = "all";
/**
 * Creates a span processor with Traceloop's custom span handling logic.
 * This can be used independently of the full SDK initialization.
 *
 * @param options - Configuration options for the span processor
 * @returns A configured SpanProcessor instance
 */
const createSpanProcessor = (options) => {
    var _a;
    const url = `${options.baseUrl || process.env.TRACELOOP_BASE_URL || "https://api.traceloop.com"}/v1/traces`;
    const headers = options.headers ||
        (process.env.TRACELOOP_HEADERS
            ? parseKeyPairsIntoRecord(process.env.TRACELOOP_HEADERS)
            : { Authorization: `Bearer ${options.apiKey}` });
    const traceExporter = (_a = options.exporter) !== null && _a !== void 0 ? _a : new OTLPTraceExporter({
        url,
        headers,
    });
    const spanProcessor = options.disableBatch
        ? new SimpleSpanProcessor(traceExporter)
        : new BatchSpanProcessor(traceExporter);
    // Store the original onEnd method
    const originalOnEnd = spanProcessor.onEnd.bind(spanProcessor);
    spanProcessor.onStart = onSpanStart;
    if (options.allowedInstrumentationLibraries === ALL_INSTRUMENTATION_LIBRARIES) {
        spanProcessor.onEnd = onSpanEnd(originalOnEnd);
    }
    else {
        const instrumentationLibraries = [...traceloopInstrumentationLibraries];
        if (options.allowedInstrumentationLibraries) {
            instrumentationLibraries.push(...options.allowedInstrumentationLibraries);
        }
        spanProcessor.onEnd = onSpanEnd(originalOnEnd, instrumentationLibraries);
    }
    return spanProcessor;
};
const traceloopInstrumentationLibraries = [
    "ai",
    "@traceloop/node-server-sdk",
    "@traceloop/instrumentation-openai",
    "@traceloop/instrumentation-langchain",
    "@traceloop/instrumentation-chroma",
    "@traceloop/instrumentation-anthropic",
    "@traceloop/instrumentation-llamaindex",
    "@traceloop/instrumentation-vertexai",
    "@traceloop/instrumentation-bedrock",
    "@traceloop/instrumentation-cohere",
    "@traceloop/instrumentation-pinecone",
    "@traceloop/instrumentation-qdrant",
    "@traceloop/instrumentation-together",
];
/**
 * Handles span start event, enriching it with workflow and entity information
 */
const onSpanStart = (span) => {
    const workflowName = context.active().getValue(WORKFLOW_NAME_KEY);
    if (workflowName) {
        span.setAttribute(SpanAttributes.TRACELOOP_WORKFLOW_NAME, workflowName);
    }
    const entityName = context.active().getValue(ENTITY_NAME_KEY);
    if (entityName) {
        span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_PATH, entityName);
    }
    const associationProperties = context
        .active()
        .getValue(ASSOCATION_PROPERTIES_KEY);
    if (associationProperties) {
        for (const [key, value] of Object.entries(associationProperties)) {
            span.setAttribute(`${SpanAttributes.TRACELOOP_ASSOCIATION_PROPERTIES}.${key}`, value);
        }
    }
};
/**
 * Handles span end event, adapting attributes for Vercel AI compatibility
 */
const onSpanEnd = (originalOnEnd, instrumentationLibraries) => {
    return (span) => {
        var _a;
        if (instrumentationLibraries &&
            !instrumentationLibraries.includes((_a = span.instrumentationScope) === null || _a === void 0 ? void 0 : _a.name)) {
            return;
        }
        // Apply AI SDK transformations (if needed)
        transformAiSdkSpan(span);
        originalOnEnd(span);
    };
};

let _sdk;
let _spanProcessor;
let openAIInstrumentation;
let anthropicInstrumentation;
let cohereInstrumentation;
let vertexaiInstrumentation;
let aiplatformInstrumentation;
let bedrockInstrumentation;
let langchainInstrumentation;
let llamaIndexInstrumentation;
let pineconeInstrumentation;
let chromadbInstrumentation;
let qdrantInstrumentation;
let togetherInstrumentation;
const instrumentations = [];
const initInstrumentations = () => {
    const exceptionLogger = (e) => Telemetry.getInstance().logException(e);
    const enrichTokens = (process.env.TRACELOOP_ENRICH_TOKENS || "true").toLowerCase() === "true";
    openAIInstrumentation = new OpenAIInstrumentation({
        enrichTokens,
        exceptionLogger,
    });
    instrumentations.push(openAIInstrumentation);
    anthropicInstrumentation = new AnthropicInstrumentation({ exceptionLogger });
    instrumentations.push(anthropicInstrumentation);
    cohereInstrumentation = new CohereInstrumentation({ exceptionLogger });
    instrumentations.push(cohereInstrumentation);
    vertexaiInstrumentation = new VertexAIInstrumentation({
        exceptionLogger,
    });
    instrumentations.push(vertexaiInstrumentation);
    aiplatformInstrumentation = new AIPlatformInstrumentation({
        exceptionLogger,
    });
    instrumentations.push(aiplatformInstrumentation);
    bedrockInstrumentation = new BedrockInstrumentation({ exceptionLogger });
    instrumentations.push(bedrockInstrumentation);
    pineconeInstrumentation = new PineconeInstrumentation({ exceptionLogger });
    instrumentations.push(pineconeInstrumentation);
    langchainInstrumentation = new LangChainInstrumentation({ exceptionLogger });
    instrumentations.push(langchainInstrumentation);
    llamaIndexInstrumentation = new LlamaIndexInstrumentation({
        exceptionLogger,
    });
    instrumentations.push(llamaIndexInstrumentation);
    chromadbInstrumentation = new ChromaDBInstrumentation({ exceptionLogger });
    instrumentations.push(chromadbInstrumentation);
    qdrantInstrumentation = new QdrantInstrumentation({ exceptionLogger });
    instrumentations.push(qdrantInstrumentation);
    togetherInstrumentation = new TogetherInstrumentation({ exceptionLogger });
    instrumentations.push(togetherInstrumentation);
};
const manuallyInitInstrumentations = (instrumentModules) => {
    const exceptionLogger = (e) => Telemetry.getInstance().logException(e);
    const enrichTokens = (process.env.TRACELOOP_ENRICH_TOKENS || "true").toLowerCase() === "true";
    // Clear the instrumentations array that was initialized by default
    instrumentations.length = 0;
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.openAI) {
        openAIInstrumentation = new OpenAIInstrumentation({
            enrichTokens,
            exceptionLogger,
        });
        instrumentations.push(openAIInstrumentation);
        openAIInstrumentation.manuallyInstrument(instrumentModules.openAI);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.anthropic) {
        anthropicInstrumentation = new AnthropicInstrumentation({
            exceptionLogger,
        });
        instrumentations.push(anthropicInstrumentation);
        anthropicInstrumentation.manuallyInstrument(instrumentModules.anthropic);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.cohere) {
        cohereInstrumentation = new CohereInstrumentation({ exceptionLogger });
        instrumentations.push(cohereInstrumentation);
        cohereInstrumentation.manuallyInstrument(instrumentModules.cohere);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.google_vertexai) {
        vertexaiInstrumentation = new VertexAIInstrumentation({
            exceptionLogger,
        });
        instrumentations.push(vertexaiInstrumentation);
        vertexaiInstrumentation.manuallyInstrument(instrumentModules.google_vertexai);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.google_aiplatform) {
        aiplatformInstrumentation = new AIPlatformInstrumentation({
            exceptionLogger,
        });
        instrumentations.push(aiplatformInstrumentation);
        aiplatformInstrumentation.manuallyInstrument(instrumentModules.google_aiplatform);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.bedrock) {
        bedrockInstrumentation = new BedrockInstrumentation({ exceptionLogger });
        instrumentations.push(bedrockInstrumentation);
        bedrockInstrumentation.manuallyInstrument(instrumentModules.bedrock);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.pinecone) {
        const instrumentation = new PineconeInstrumentation({ exceptionLogger });
        instrumentations.push(instrumentation);
        instrumentation.manuallyInstrument(instrumentModules.pinecone);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.langchain) {
        langchainInstrumentation = new LangChainInstrumentation({
            exceptionLogger,
        });
        instrumentations.push(langchainInstrumentation);
        langchainInstrumentation.manuallyInstrument(instrumentModules.langchain);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.llamaIndex) {
        llamaIndexInstrumentation = new LlamaIndexInstrumentation({
            exceptionLogger,
        });
        instrumentations.push(llamaIndexInstrumentation);
        llamaIndexInstrumentation.manuallyInstrument(instrumentModules.llamaIndex);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.chromadb) {
        chromadbInstrumentation = new ChromaDBInstrumentation({ exceptionLogger });
        instrumentations.push(chromadbInstrumentation);
        chromadbInstrumentation.manuallyInstrument(instrumentModules.chromadb);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.qdrant) {
        qdrantInstrumentation = new QdrantInstrumentation({ exceptionLogger });
        instrumentations.push(qdrantInstrumentation);
        qdrantInstrumentation.manuallyInstrument(instrumentModules.qdrant);
    }
    if (instrumentModules === null || instrumentModules === void 0 ? void 0 : instrumentModules.together) {
        togetherInstrumentation = new TogetherInstrumentation({ exceptionLogger });
        instrumentations.push(togetherInstrumentation);
        togetherInstrumentation.manuallyInstrument(instrumentModules.together);
    }
};
/**
 * Initializes the Traceloop SDK.
 * Must be called once before any other SDK methods.
 *
 * @param options - The options to initialize the SDK. See the {@link InitializeOptions} for details.
 * @throws {InitializationError} if the configuration is invalid or if failed to fetch feature data.
 */
const startTracing = (options) => {
    var _a;
    if (Object.keys(options.instrumentModules || {}).length > 0) {
        manuallyInitInstrumentations(options.instrumentModules);
    }
    if (!shouldSendTraces()) {
        openAIInstrumentation === null || openAIInstrumentation === void 0 ? void 0 : openAIInstrumentation.setConfig({
            traceContent: false,
        });
        llamaIndexInstrumentation === null || llamaIndexInstrumentation === void 0 ? void 0 : llamaIndexInstrumentation.setConfig({
            traceContent: false,
        });
        vertexaiInstrumentation === null || vertexaiInstrumentation === void 0 ? void 0 : vertexaiInstrumentation.setConfig({
            traceContent: false,
        });
        aiplatformInstrumentation === null || aiplatformInstrumentation === void 0 ? void 0 : aiplatformInstrumentation.setConfig({
            traceContent: false,
        });
        bedrockInstrumentation === null || bedrockInstrumentation === void 0 ? void 0 : bedrockInstrumentation.setConfig({
            traceContent: false,
        });
        cohereInstrumentation === null || cohereInstrumentation === void 0 ? void 0 : cohereInstrumentation.setConfig({
            traceContent: false,
        });
        chromadbInstrumentation === null || chromadbInstrumentation === void 0 ? void 0 : chromadbInstrumentation.setConfig({
            traceContent: false,
        });
        togetherInstrumentation === null || togetherInstrumentation === void 0 ? void 0 : togetherInstrumentation.setConfig({
            traceContent: false,
        });
    }
    const headers = options.headers ||
        (process.env.TRACELOOP_HEADERS
            ? parseKeyPairsIntoRecord(process.env.TRACELOOP_HEADERS)
            : { Authorization: `Bearer ${options.apiKey}` });
    const traceExporter = (_a = options.exporter) !== null && _a !== void 0 ? _a : new OTLPTraceExporter({
        url: `${options.baseUrl}/v1/traces`,
        headers,
    });
    _spanProcessor = createSpanProcessor({
        apiKey: options.apiKey,
        baseUrl: options.baseUrl,
        disableBatch: options.disableBatch,
        exporter: traceExporter,
        headers,
        allowedInstrumentationLibraries: ALL_INSTRUMENTATION_LIBRARIES,
    });
    const spanProcessors = [_spanProcessor];
    if (options.processor) {
        spanProcessors.push(options.processor);
    }
    _sdk = new NodeSDK({
        resource: resourceFromAttributes({
            [ATTR_SERVICE_NAME]: options.appName || process.env.npm_package_name,
        }),
        spanProcessors,
        contextManager: options.contextManager,
        textMapPropagator: options.propagator,
        traceExporter,
        instrumentations,
        // We should re-consider removing irrelevant spans here in the future
        // sampler: new TraceloopSampler(),
    });
    _sdk.start();
};
const shouldSendTraces = () => {
    if (!_configuration) {
        diag.warn("Traceloop not initialized");
        return false;
    }
    const contextShouldSendPrompts = context
        .active()
        .getValue(CONTEXT_KEY_ALLOW_TRACE_CONTENT);
    if (contextShouldSendPrompts !== undefined) {
        return contextShouldSendPrompts;
    }
    if (_configuration.traceContent === false ||
        (process.env.TRACELOOP_TRACE_CONTENT || "true").toLowerCase() === "false") {
        return false;
    }
    return true;
};
const forceFlush = async () => {
    await _spanProcessor.forceFlush();
};

function withEntity(type, { name, version, associationProperties, traceContent: overrideTraceContent, inputParameters, suppressTracing: shouldSuppressTracing, }, fn, thisArg, ...args) {
    let entityContext = context.active();
    if (type === TraceloopSpanKindValues.WORKFLOW ||
        type === TraceloopSpanKindValues.AGENT) {
        entityContext = entityContext.setValue(WORKFLOW_NAME_KEY, name);
    }
    const entityPath = getEntityPath(entityContext);
    if (type === TraceloopSpanKindValues.TOOL ||
        type === TraceloopSpanKindValues.TASK) {
        const fullEntityName = entityPath ? `${entityPath}.${name}` : name;
        entityContext = entityContext.setValue(ENTITY_NAME_KEY, fullEntityName);
    }
    if (overrideTraceContent != undefined) {
        entityContext = entityContext.setValue(CONTEXT_KEY_ALLOW_TRACE_CONTENT, overrideTraceContent);
    }
    if (associationProperties) {
        entityContext = entityContext.setValue(ASSOCATION_PROPERTIES_KEY, associationProperties);
    }
    if (shouldSuppressTracing) {
        entityContext = suppressTracing(entityContext);
    }
    return context.with(entityContext, () => getTracer().startActiveSpan(`${name}.${type}`, {}, entityContext, async (span) => {
        if (type === TraceloopSpanKindValues.WORKFLOW ||
            type === TraceloopSpanKindValues.AGENT) {
            span.setAttribute(SpanAttributes.TRACELOOP_WORKFLOW_NAME, name);
        }
        span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_NAME, name);
        span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_PATH, entityPath || "");
        span.setAttribute(SpanAttributes.TRACELOOP_SPAN_KIND, type);
        if (version) {
            span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_VERSION, version);
        }
        if (shouldSendTraces()) {
            try {
                const input = inputParameters !== null && inputParameters !== void 0 ? inputParameters : args;
                if (input.length === 1 &&
                    typeof input[0] === "object" &&
                    !(input[0] instanceof Map)) {
                    span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_INPUT, serialize({ args: [], kwargs: input[0] }));
                }
                else {
                    span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_INPUT, serialize({
                        args: input,
                        kwargs: {},
                    }));
                }
            }
            catch (error) {
                Telemetry.getInstance().logException(error);
            }
        }
        const res = fn.apply(thisArg, args);
        if (res instanceof Promise) {
            return res.then((resolvedRes) => {
                try {
                    if (shouldSendTraces()) {
                        span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_OUTPUT, serialize(resolvedRes));
                    }
                }
                catch (error) {
                    Telemetry.getInstance().logException(error);
                }
                finally {
                    span.end();
                }
                return resolvedRes;
            });
        }
        try {
            if (shouldSendTraces()) {
                span.setAttribute(SpanAttributes.TRACELOOP_ENTITY_OUTPUT, serialize(res));
            }
        }
        catch (error) {
            Telemetry.getInstance().logException(error);
        }
        finally {
            span.end();
        }
        return res;
    }));
}
function withWorkflow(config, fn, ...args) {
    return withEntity(TraceloopSpanKindValues.WORKFLOW, config, fn, undefined, ...args);
}
function withTask(config, fn, ...args) {
    return withEntity(TraceloopSpanKindValues.TASK, config, fn, undefined, ...args);
}
function withAgent(config, fn, ...args) {
    return withEntity(TraceloopSpanKindValues.AGENT, config, fn, undefined, ...args);
}
function withTool(config, fn, ...args) {
    return withEntity(TraceloopSpanKindValues.TOOL, config, fn, undefined, ...args);
}
function entity(type, config) {
    return function (target, propertyKey, descriptor) {
        const originalMethod = descriptor.value;
        descriptor.value = function (...args) {
            var _a;
            let actualConfig;
            if (typeof config === "function") {
                actualConfig = config(this, ...args);
            }
            else {
                actualConfig = config;
            }
            const entityName = (_a = actualConfig.name) !== null && _a !== void 0 ? _a : originalMethod.name;
            return withEntity(type, Object.assign(Object.assign({}, actualConfig), { name: entityName }), originalMethod, this, ...args);
        };
    };
}
function workflow(config) {
    return entity(TraceloopSpanKindValues.WORKFLOW, config !== null && config !== void 0 ? config : {});
}
function task(config) {
    return entity(TraceloopSpanKindValues.TASK, config !== null && config !== void 0 ? config : {});
}
function agent(config) {
    return entity(TraceloopSpanKindValues.AGENT, config !== null && config !== void 0 ? config : {});
}
function tool(config) {
    return entity(TraceloopSpanKindValues.TOOL, config !== null && config !== void 0 ? config : {});
}
function cleanInput(input) {
    if (input instanceof Map) {
        return Array.from(input.entries());
    }
    else if (Array.isArray(input)) {
        return input.map((value) => cleanInput(value));
    }
    else if (!input) {
        return input;
    }
    else if (typeof input === "object") {
        // serialize object one by one
        const output = {};
        Object.entries(input).forEach(([key, value]) => {
            output[key] = cleanInput(value);
        });
        return output;
    }
    return input;
}
function serialize(input) {
    return JSON.stringify(cleanInput(input));
}

class VectorSpan {
    constructor(span) {
        this.span = span;
    }
    reportQuery({ queryVector }) {
        if (!shouldSendTraces()) {
            this.span.addEvent(Events.DB_QUERY_EMBEDDINGS);
        }
        this.span.addEvent(Events.DB_QUERY_EMBEDDINGS, {
            [EventAttributes.DB_QUERY_EMBEDDINGS_VECTOR]: JSON.stringify(queryVector),
        });
    }
    reportResults({ results, }) {
        for (let i = 0; i < results.length; i++) {
            this.span.addEvent(Events.DB_QUERY_RESULT, {
                [EventAttributes.DB_QUERY_RESULT_ID]: results[i].ids,
                [EventAttributes.DB_QUERY_RESULT_SCORE]: results[i].scores,
                [EventAttributes.DB_QUERY_RESULT_DISTANCE]: results[i].distances,
                [EventAttributes.DB_QUERY_RESULT_METADATA]: JSON.stringify(results[i].metadata),
                [EventAttributes.DB_QUERY_RESULT_VECTOR]: results[i].vectors,
                [EventAttributes.DB_QUERY_RESULT_DOCUMENT]: results[i].documents,
            });
        }
    }
}
class LLMSpan {
    constructor(span) {
        this.span = span;
    }
    reportRequest({ model, messages, }) {
        this.span.setAttributes({
            [SpanAttributes.LLM_REQUEST_MODEL]: model,
        });
        messages.forEach((message, index) => {
            this.span.setAttributes({
                [`${SpanAttributes.LLM_PROMPTS}.${index}.role`]: message.role,
                [`${SpanAttributes.LLM_PROMPTS}.${index}.content`]: typeof message.content === "string"
                    ? message.content
                    : JSON.stringify(message.content),
            });
        });
    }
    reportResponse({ model, usage, completions, }) {
        this.span.setAttribute(SpanAttributes.LLM_RESPONSE_MODEL, model);
        if (usage) {
            this.span.setAttributes({
                [SpanAttributes.LLM_USAGE_PROMPT_TOKENS]: usage.prompt_tokens,
                [SpanAttributes.LLM_USAGE_COMPLETION_TOKENS]: usage.completion_tokens,
                [SpanAttributes.LLM_USAGE_TOTAL_TOKENS]: usage.total_tokens,
            });
        }
        completions === null || completions === void 0 ? void 0 : completions.forEach((completion, index) => {
            this.span.setAttributes({
                [`${SpanAttributes.LLM_COMPLETIONS}.${index}.finish_reason`]: completion.finish_reason,
                [`${SpanAttributes.LLM_COMPLETIONS}.${index}.role`]: completion.message.role,
                [`${SpanAttributes.LLM_COMPLETIONS}.${index}.content`]: completion.message.content || "",
            });
        });
    }
}
function withVectorDBCall({ vendor, type }, fn, thisArg) {
    const entityContext = context.active();
    return getTracer().startActiveSpan(`${vendor}.${type}`, { [SpanAttributes.LLM_REQUEST_TYPE]: type }, entityContext, (span) => {
        const res = fn.apply(thisArg, [{ span: new VectorSpan(span) }]);
        if (res instanceof Promise) {
            return res.then((resolvedRes) => {
                span.end();
                return resolvedRes;
            });
        }
        span.end();
        return res;
    });
}
function withLLMCall({ vendor, type }, fn, thisArg) {
    const span = getTracer().startSpan(`${vendor}.${type}`, {}, context.active());
    span.setAttribute(SpanAttributes.LLM_REQUEST_TYPE, type);
    trace.setSpan(context.active(), span);
    const res = fn.apply(thisArg, [{ span: new LLMSpan(span) }]);
    if (res instanceof Promise) {
        return res.then((resolvedRes) => {
            span.end();
            return resolvedRes;
        });
    }
    span.end();
    return res;
}

function withAssociationProperties(properties, fn, thisArg, ...args) {
    if (Object.keys(properties).length === 0) {
        return fn.apply(thisArg, args);
    }
    const newContext = context
        .active()
        .setValue(ASSOCATION_PROPERTIES_KEY, properties);
    return context.with(newContext, fn, thisArg, ...args);
}

/**
 * Reports a custom metric to the current active span.
 *
 * This function allows you to add a custom metric to the current span in the trace.
 * If there is no active span, a warning will be logged.
 *
 * @param {string} metricName - The name of the custom metric.
 * @param {number} metricValue - The numeric value of the custom metric.
 *
 * @example
 * reportCustomMetric('processing_time', 150);
 */
const reportCustomMetric = (metricName, metricValue) => {
    const currentContext = context.active();
    const currentSpan = trace.getSpan(currentContext);
    if (currentSpan) {
        currentSpan.setAttribute(`traceloop.custom_metric.${metricName}`, metricValue);
    }
    else {
        diag.warn(`No active span found to report custom metric: ${metricName}`);
    }
};

const TEMPLATING_ENGINE = {
    JINJA2: "jinja2",
};

const env = new Environment(null, {
    throwOnUndefined: true, // throw error if param not found
});
const renderMessages = (promptVersion, variables) => {
    if (promptVersion.templating_engine === TEMPLATING_ENGINE.JINJA2) {
        return promptVersion.messages.map((message) => {
            try {
                if (typeof message.template === "string") {
                    return {
                        content: env.renderString(message.template, variables),
                        role: message.role,
                    };
                }
                else {
                    return {
                        content: message.template.map((content) => {
                            if (content.type === "text") {
                                return {
                                    type: "text",
                                    text: env.renderString(content.text, variables),
                                };
                            }
                            else {
                                return content;
                            }
                        }),
                        role: message.role,
                    };
                }
            }
            catch (err) {
                throw new TraceloopError(`Failed to render message template. Missing variables?`);
            }
        });
    }
    else {
        throw new TraceloopError(`Templating engine ${promptVersion.templating_engine} is not supported`);
    }
};

const getEffectiveVersion = (prompt) => {
    const version = prompt.versions.find((v) => v.id === prompt.target.version);
    if (!version) {
        throw new TraceloopError(`Prompt version ${prompt.target.version} not found`);
    }
    return version;
};
const managedPromptTracingAttributes = (prompt, promptVersion, variables) => {
    const variableAttributes = Object.keys(variables).reduce((acc, key) => {
        acc[`traceloop.prompt.template_variables.${key}`] = variables[key];
        return acc;
    }, {});
    return Object.assign({ "traceloop.prompt.key": prompt.key, "traceloop.prompt.version": promptVersion.version, "traceloop.prompt.version_hash": promptVersion.hash, "traceloop.prompt.version_name": promptVersion.name }, variableAttributes);
};
const getPrompt = (key, variables) => {
    var _a;
    Telemetry.getInstance().capture("prompt:rendered");
    const prompt = getPromptByKey(key);
    const promptVersion = getEffectiveVersion(prompt);
    let result = {}; //TODO - SDK needs to do work specific to each vendor/model? maybe we do this in the backend?
    if (promptVersion.llm_config.mode === "completion") {
        const message = renderMessages(promptVersion, variables);
        result = Object.assign(Object.assign({}, promptVersion.llm_config), { prompt: (_a = message === null || message === void 0 ? void 0 : message[0]) === null || _a === void 0 ? void 0 : _a.content });
    }
    else {
        result = Object.assign({ messages: renderMessages(promptVersion, variables) }, promptVersion.llm_config);
    }
    if ((result === null || result === void 0 ? void 0 : result["stop"].length) === 0)
        delete result["stop"];
    delete result["mode"];
    result.extraAttributes = managedPromptTracingAttributes(prompt, promptVersion, variables);
    return result;
};

initInstrumentations();

export { ALL_INSTRUMENTATION_LIBRARIES, ArgumentNotProvidedError, InitializationError, LLMSpan, NotInitializedError, PromptNotFoundError, SEVERITY, TraceloopClient, TraceloopError, VectorSpan, agent, createSpanProcessor, forceFlush, getClient, getPrompt, initialize, reportCustomMetric, task, tool, traceloopInstrumentationLibraries, waitForInitialization, withAgent, withAssociationProperties, withLLMCall, withTask, withTool, withVectorDBCall, withWorkflow, workflow };
