// @nevermined-io/payments/dist/mcp/index.js
import { createOAuthRouter } from './http/oauth-router.js';
import { createMcpApp, startManagedServer } from './http/managed-server.js';
import { PaywallDecorator } from './core/paywall.js';
import { PaywallAuthenticator } from './core/auth.js';
import { CreditsContextProvider } from './core/credits-context.js';
import { createServerManager } from './core/server-manager.js';
// Re-export utilities
export { buildExtraFromHttpHeaders, buildExtraFromHttpRequest } from './utils/extra.js';
// Re-export HTTP module components for advanced usage
export { getOAuthUrls, buildProtectedResourceMetadata, buildMcpProtectedResourceMetadata, buildAuthorizationServerMetadata, buildOidcConfiguration, buildServerInfoResponse, createOAuthRouter, createCorsMiddleware, createJsonMiddleware, createRequireAuthMiddleware, createHttpLoggingMiddleware, startManagedServer, createMcpApp, ClientRegistrationError, } from './http/index.js';
/**
 * Build MCP integration with modular architecture.
 * This function creates the complete MCP API surface including
 * paywall protection, OAuth endpoints, and HTTP server management.
 *
 * @param paymentsService - The Payments instance
 * @returns MCP integration API
 */
export function buildMcpIntegration(paymentsService) {
    const authenticator = new PaywallAuthenticator(paymentsService);
    const creditsContext = new CreditsContextProvider();
    const paywallDecorator = new PaywallDecorator(paymentsService, authenticator, creditsContext);
    // Extended configuration storage
    let extendedConfig = {
        agentId: '',
        serverName: 'mcp-server',
    };
    /**
     * Configure the MCP integration.
     * This sets up the agent ID, server name, and optional HTTP settings.
     *
     * @param options - Configuration options
     *
     * @example
     * ```typescript
     * payments.mcp.configure({
     *   agentId: 'agent_123',
     *   serverName: 'my-mcp-server',
     *   baseUrl: 'http://localhost:5001',
     *   environment: 'staging_sandbox',
     *   tools: ['hello_world', 'weather']
     * })
     * ```
     */
    function configure(options) {
        extendedConfig = {
            ...extendedConfig,
            ...options,
        };
        // Also configure the paywall decorator
        paywallDecorator.configure({
            agentId: options.agentId,
            serverName: options.serverName,
        });
    }
    function withPaywall(handler, options) {
        const opts = options ?? { kind: 'tool', name: 'unnamed' };
        return paywallDecorator.protect(handler, opts);
    }
    /**
     * Authenticate meta MCP operations (initialize, tools/list, etc.).
     *
     * @param extra - The extra parameter from MCP
     * @param method - The MCP method being called
     * @returns Authentication result
     */
    async function authenticateMeta(extra, options = {}, method) {
        const cfg = extendedConfig;
        const agentId = cfg.agentId || '';
        const serverName = cfg.serverName || 'mcp-server';
        return authenticator.authenticateMeta(extra, { planId: options?.planId }, agentId, serverName, method);
    }
    /**
     * Attach paywall protection to an MCP server.
     * Returns a registrar with protected register methods.
     *
     * @param server - The MCP server instance
     * @returns Protected registrar
     */
    function attach(server) {
        return {
            registerTool(name, config, handler, options) {
                const protectedHandler = withPaywall(handler, { kind: 'tool', name, ...(options || {}) });
                server.registerTool(name, config, protectedHandler);
            },
            registerResource(name, template, config, handler, options) {
                const protectedHandler = withPaywall(handler, {
                    kind: 'resource',
                    name,
                    ...(options || {}),
                });
                server.registerResource(name, template, config, protectedHandler);
            },
            registerPrompt(name, config, handler, options) {
                const protectedHandler = withPaywall(handler, { kind: 'prompt', name, ...(options || {}) });
                server.registerPrompt(name, config, protectedHandler);
            },
        };
    }
    /**
     * Create an Express router with OAuth 2.1 endpoints.
     * Use this when you already have an Express app and want to add OAuth support.
     *
     * @param options - Router configuration
     * @returns Express Router
     *
     * @example
     * ```typescript
     * const router = payments.mcp.createRouter({
     *   baseUrl: 'http://localhost:5001',
     *   serverName: 'my-mcp-server',
     *   tools: ['hello_world']
     * })
     *
     * app.use(router)
     * ```
     */
    function createRouter(options) {
        const agentId = options.agentId || extendedConfig.agentId;
        if (!agentId) {
            throw new Error('agentId is required. Either pass it in options or call configure() first.');
        }
        const environment = extendedConfig.environment || paymentsService.environmentName || 'staging_sandbox';
        return createOAuthRouter({
            payments: paymentsService,
            baseUrl: options.baseUrl,
            agentId,
            environment,
            serverName: options.serverName || extendedConfig.serverName || 'mcp-server',
            tools: options.tools || extendedConfig.tools || [],
            scopes: extendedConfig.scopes,
            enableOAuthDiscovery: options.enableOAuthDiscovery ?? true,
            enableClientRegistration: options.enableClientRegistration ?? true,
            enableHealthCheck: options.enableHealthCheck ?? true,
            enableServerInfo: options.enableServerInfo ?? true,
            version: options.version,
            description: options.description,
        });
    }
    /**
     * Create an Express app pre-configured with OAuth discovery endpoints.
     * Use this when you want more control over the app before starting.
     *
     * Note: Authentication is handled by withPaywall() on each tool,
     * not by HTTP middleware.
     *
     * @param options - App configuration
     * @returns Configured Express app with OAuth endpoints
     *
     * @example
     * ```typescript
     * const app = payments.mcp.createApp({
     *   baseUrl: 'http://localhost:5001',
     *   serverName: 'my-mcp-server'
     * })
     *
     * // Add your MCP handler - auth via withPaywall() on each tool
     * app.post('/mcp', mcpHandler)
     *
     * app.listen(5001)
     * ```
     */
    function createApp(options) {
        const agentId = options.agentId || extendedConfig.agentId;
        if (!agentId) {
            throw new Error('agentId is required. Either pass it in options or call configure() first.');
        }
        const environment = extendedConfig.environment || paymentsService.environmentName || 'staging_sandbox';
        return createMcpApp({
            payments: paymentsService,
            baseUrl: options.baseUrl,
            agentId,
            environment,
            serverName: options.serverName || extendedConfig.serverName || 'mcp-server',
            tools: options.tools || extendedConfig.tools || [],
            scopes: extendedConfig.scopes,
            enableOAuthDiscovery: options.enableOAuthDiscovery ?? true,
            enableClientRegistration: options.enableClientRegistration ?? true,
            enableHealthCheck: options.enableHealthCheck ?? true,
            enableServerInfo: options.enableServerInfo ?? true,
            version: options.version,
            description: options.description,
        });
    }
    /**
     * Start a managed HTTP server with OAuth 2.1 support.
     * This creates a complete Express server with all OAuth endpoints pre-configured.
     *
     * @param options - Server configuration
     * @returns Server result with control methods
     *
     * @example
     * ```typescript
     * const { baseUrl, stop } = await payments.mcp.startServer({
     *   port: 5001,
     *   serverName: 'my-mcp-server',
     *   tools: ['hello_world', 'weather']
     * })
     *
     * console.log(`Server running at ${baseUrl}`)
     *
     * // Later: gracefully stop
     * await stop()
     * ```
     */
    async function startServer(options) {
        const agentId = options.agentId || extendedConfig.agentId;
        if (!agentId) {
            throw new Error('agentId is required. Either pass it in options or call configure() first.');
        }
        const baseUrl = options.baseUrl || `http://localhost:${options.port}`;
        const environment = extendedConfig.environment || paymentsService.environmentName || 'staging_sandbox';
        return startManagedServer({
            payments: paymentsService,
            port: options.port,
            host: options.host,
            baseUrl,
            agentId,
            environment,
            serverName: options.serverName || extendedConfig.serverName || 'mcp-server',
            tools: options.tools || extendedConfig.tools || [],
            scopes: extendedConfig.scopes,
            enableOAuthDiscovery: options.enableOAuthDiscovery ?? true,
            enableClientRegistration: options.enableClientRegistration ?? true,
            enableHealthCheck: options.enableHealthCheck ?? true,
            enableServerInfo: options.enableServerInfo ?? true,
            version: options.version,
            description: options.description,
            onStart: options.onStart,
            onLog: options.onLog,
        });
    }
    /**
     * Get the current configuration.
     *
     * @returns Current MCP configuration
     */
    function getConfig() {
        return { ...extendedConfig };
    }
    // =============================================================================
    // SIMPLIFIED API - High-level API that hides McpServer, Transport, and Express
    // =============================================================================
    // Server manager for simplified API
    let serverManager = null;
    /**
     * Get or create the server manager.
     */
    function getServerManager() {
        if (!serverManager) {
            serverManager = createServerManager(paymentsService);
        }
        return serverManager;
    }
    /**
     * Register a tool with the simplified API.
     * Must be called before start().
     *
     * @param name - Tool name
     * @param config - Tool configuration
     * @param handler - Tool handler function
     * @param options - Registration options (credits, etc.)
     *
     * @example
     * ```typescript
     * payments.mcp.registerTool(
     *   'hello_world',
     *   {
     *     description: 'Returns a hello world message',
     *     inputSchema: { name: { type: 'string' } }
     *   },
     *   async (args: { name: string }, context?: ToolContext) => ({
     *     content: [{ type: 'text', text: `Hello, ${args.name}!` }]
     *   }),
     *   { credits: 1, onRedeemError: 'ignore' }
     * )
     * ```
     */
    function registerTool(name, config, handler, options) {
        getServerManager().registerTool(name, config, handler, options);
    }
    /**
     * Register a resource with the simplified API.
     * Must be called before start().
     *
     * @param name - Resource name
     * @param uriOrTemplate - Resource URI (string) or template
     * @param config - Resource metadata configuration
     * @param handler - Resource handler function
     * @param options - Registration options (credits, etc.)
     */
    function registerResource(name, uriOrTemplate, config, handler, options) {
        getServerManager().registerResource(name, uriOrTemplate, config, handler, options);
    }
    /**
     * Register a prompt with the simplified API.
     * Must be called before start().
     * Matches the signature of MCP SDK registerPrompt.
     *
     * @param name - Prompt name
     * @param config - Prompt configuration
     * @param handler - Prompt handler function
     * @param options - Registration options (credits, etc.)
     */
    function registerPrompt(name, config, handler, options) {
        getServerManager().registerPrompt(name, config, handler, options);
    }
    /**
     * Start the MCP server with the simplified API.
     * This creates and starts everything: McpServer, Express, OAuth endpoints, etc.
     *
     * @param config - Server configuration
     * @returns Server result with stop() method
     *
     * @example
     * ```typescript
     * // Register tools first
     * payments.mcp.registerTool('hello', { description: '...' }, handler)
     *
     * // Then start the server
     * const { info, stop } = await payments.mcp.start({
     *   port: 5001,
     *   agentId: 'did:nv:...',
     *   serverName: 'my-mcp-server'
     * })
     *
     * console.log(`Server running at ${info.baseUrl}`)
     *
     * // Later: stop gracefully
     * await stop()
     * ```
     */
    async function start(config) {
        return getServerManager().start(config);
    }
    /**
     * Stop the MCP server.
     * This is a convenience method - you can also use the stop() from start()'s result.
     */
    async function stop() {
        if (serverManager) {
            await serverManager.stop();
        }
    }
    return {
        // Core paywall functionality (advanced API)
        configure,
        withPaywall,
        attach,
        authenticateMeta,
        // HTTP/OAuth functionality (advanced API)
        createRouter,
        createApp,
        startServer,
        // Utilities
        getConfig,
        // Simplified API (recommended for most users)
        registerTool,
        registerResource,
        registerPrompt,
        start,
        stop,
    };
}
