// @nevermined-io/payments/dist/payments.js
import { AIQueryApi } from './api/query-api.js';
import { PaymentsError } from './common/payments.error.js';
import { BasePaymentsAPI } from './api/base-payments.js';
import { PlansAPI } from './api/plans-api.js';
import { ContractsAPI } from './api/contracts-api.js';
import { AgentsAPI } from './api/agents-api.js';
import { AgentRequestsAPI } from './api/requests-api.js';
import { ObservabilityAPI } from './api/observability-api/observability-api.js';
import { ClientRegistry } from './a2a/clientRegistry.js';
import { PaymentsA2AServer } from './a2a/server.js';
import { buildPaymentAgentCard } from './a2a/agent-card.js';
import * as mcpModule from './mcp/index.js';
import { OrganizationsAPI } from './api/organizations-api/organizations-api.js';
import { FacilitatorAPI } from './x402/facilitator-api.js';
import { X402TokenAPI } from './x402/token.js';
import { DelegationAPI } from './x402/delegation-api.js';
import { VisaFacilitatorAPI } from './x402/visa-facilitator-api.js';
import { VisaTokenAPI } from './x402/visa-token-api.js';
/**
 * Main class that interacts with the Nevermined payments API.
 * Use `Payments.getInstance` for server-side usage or `Payments.getBrowserInstance` for browser usage.
 * @remarks This API requires a Nevermined API Key, which can be obtained by logging in to the Nevermined App.
 *
 * The library provides methods to manage AI Agents, Plans & process AI Agent Requests.
 *
 * Each of these functionalities is encapsulated in its own API class:
 * - `plans`: Manages AI Plans, including registration and ordering and retrieving plan details.
 * - `agents`: Handles AI Agents, including registration of AI Agents and access token generation.
 * - `requests`: Manages requests received by AI Agents, including validation and tracking.
 * - `observability`: Provides observability and logging utilities for AI Agents with Helicone integration
 */
export class Payments extends BasePaymentsAPI {
    /**
     * Exposes A2A server and client registry methods.
     * The client registry is initialized only if getClient is called.
     */
    get a2a() {
        return {
            /**
             * Starts the A2A server with payment integration.
             * @param options - Server options.
             */
            start: (options) => PaymentsA2AServer.start({ ...options, paymentsService: this }),
            /**
             * Gets (or creates) a RegisteredPaymentsClient for the given alias.
             * The registry is initialized only on first use.
             * @param options - ClientRegistryOptions.
             */
            getClient: async (options) => {
                if (!this._a2aRegistry) {
                    this._a2aRegistry = new ClientRegistry(this);
                }
                return await this._a2aRegistry.getClient(options);
            },
        };
    }
    /**
     * Returns the MCP integration API. The instance is memoized so that configuration
     * set via `configure({ agentId, serverName })` persists across calls.
     */
    get mcp() {
        if (!this._mcpIntegration) {
            this._mcpIntegration = mcpModule.buildMcpIntegration(this);
        }
        return this._mcpIntegration;
    }
    /**
     * Returns the Delegation API for listing enrolled payment methods.
     * The instance is lazily initialized on first access.
     */
    get delegation() {
        if (!this._delegation) {
            this._delegation = DelegationAPI.getInstance({
                nvmApiKey: this.nvmApiKey,
                environment: this.environmentName,
            });
        }
        return this._delegation;
    }
    /**
     * Get an instance of the Payments class for server-side usage.
     *
     * @param options - The options to initialize the payments class.
     * @example
     * ```
     * const payments = Payments.getInstance({
     *   nvmApiKey: 'your-nvm-api-key',
     *   environment: 'sandbox'
     * })
     * ```
     * @returns An instance of {@link Payments}
     * @throws PaymentsError if nvmApiKey is missing.
     */
    static getInstance(options) {
        if (!options.nvmApiKey) {
            throw new PaymentsError('Nevermined API Key is required');
        }
        return new Payments(options, false);
    }
    /**
     * Get an instance of the Payments class for browser usage.
     *
     * @remarks
     * This is a browser-only function.
     *
     * @param options - The options to initialize the payments class.
     * @example
     * ```
     * const payments = Payments.getBrowserInstance({
     *   returnUrl: 'https://mysite.example',
     *   environment: 'sandbox',
     *   appId: 'my-app-id',
     *   version: '1.0.0'
     * })
     * ```
     * @returns An instance of {@link Payments}
     * @throws PaymentsError if returnUrl is missing.
     */
    static getBrowserInstance(options) {
        if (!options.returnUrl) {
            throw new PaymentsError('returnUrl is required');
        }
        const url = new URL(window.location.href);
        const urlNvmApiKey = url.searchParams.get('nvmApiKey');
        if (urlNvmApiKey) {
            url.searchParams.delete('nvmApiKey');
        }
        const urlAccountAddress = url.searchParams.get('accountAddress');
        if (urlAccountAddress) {
            url.searchParams.delete('accountAddress');
        }
        history.replaceState(history.state, '', url.toString());
        return new Payments(options, true);
    }
    /**
     * Initializes the Payments class.
     *
     * @param options - The options to initialize the payments class.
     * @param isBrowserInstance - Whether this instance is for browser usage.
     */
    constructor(options, isBrowserInstance = true) {
        super(options);
        this.isBrowserInstance = isBrowserInstance;
        this.initializeApi(options);
    }
    /**
     * Initializes the AI Query Protocol API.
     */
    initializeApi(options) {
        this.plans = PlansAPI.getInstance(options);
        this.agents = AgentsAPI.getInstance(options);
        this.requests = AgentRequestsAPI.getInstance(options);
        this.observability = ObservabilityAPI.getInstance(options);
        this.organizations = OrganizationsAPI.getInstance(options);
        this.query = AIQueryApi.getInstance();
        this.contracts = new ContractsAPI(options);
        if (options.scheme === 'visa') {
            this.facilitator = VisaFacilitatorAPI.getInstance(options);
            this.x402 = VisaTokenAPI.getInstance(options);
        }
        else {
            this.facilitator = FacilitatorAPI.getInstance(options);
            this.x402 = X402TokenAPI.getInstance(options);
        }
    }
    /**
     * Initiates the connect flow. The user's browser will be redirected to
     * the Nevermined App login page.
     *
     * @remarks
     * This is a browser-only function.
     * @example
     * ```
     * payments.connect()
     * ```
     */
    connect() {
        if (!this.isBrowserInstance)
            return;
        const url = new URL(`/login?returnUrl=${this.returnUrl}`, this.environment.frontend);
        window.location.href = url.toString();
    }
    /**
     * Logs out the user by removing the NVM API key.
     *
     * @remarks
     * This is a browser-only function.
     * @example
     * ```
     * payments.logout()
     * ```
     */
    logout() {
        this.nvmApiKey = '';
    }
    /**
     * Checks if a user is logged in.
     * @example
     * ```
     * payments.isLoggedIn
     * ```
     * @returns True if the user is logged in.
     */
    get isLoggedIn() {
        return this.nvmApiKey.length > 0;
    }
}
/**
 * Static A2A helpers and utilities.
 * Example: Payments.a2a.buildPaymentAgentCard(...)
 */
Payments.a2a = { buildPaymentAgentCard };
