// @nevermined-io/payments/dist/plans.js
import { PlanRedemptionType } from './common/types.js';
import { ZeroAddress } from './environments.js';
import { isEthereumAddress } from './utils.js';
export const ONE_DAY_DURATION = 86400n; // 24 * 60 * 60 seconds
export const ONE_WEEK_DURATION = 604800n; // 7 * 24 * 60 * 60 seconds
export const ONE_MONTH_DURATION = 2629746n; // (365.25 days/year ÷ 12 months/year) × 24 × 60 × 60 ≈ 2,629,746 seconds
export const ONE_YEAR_DURATION = 31557600n; // 365.25 * 24 * 60 * 60 seconds
export const getFiatPriceConfig = (amount, receiver) => {
    if (!isEthereumAddress(receiver))
        throw new Error(`Receiver address ${receiver} is not a valid Ethereum address`);
    return {
        tokenAddress: ZeroAddress,
        amounts: [amount],
        receivers: [receiver],
        contractAddress: ZeroAddress,
        feeController: ZeroAddress,
        externalPriceAddress: ZeroAddress,
        templateAddress: ZeroAddress,
        isCrypto: false,
    };
};
export const getCryptoPriceConfig = (amount, receiver, tokenAddress = ZeroAddress) => {
    if (!isEthereumAddress(receiver))
        throw new Error(`Receiver address ${receiver} is not a valid Ethereum address`);
    return {
        tokenAddress,
        amounts: [amount],
        receivers: [receiver],
        contractAddress: ZeroAddress,
        feeController: ZeroAddress,
        externalPriceAddress: ZeroAddress,
        templateAddress: ZeroAddress,
        isCrypto: true,
    };
};
export const getERC20PriceConfig = (amount, tokenAddress, receiver) => {
    return getCryptoPriceConfig(amount, receiver, tokenAddress);
};
export const getFreePriceConfig = () => {
    return {
        tokenAddress: ZeroAddress,
        amounts: [],
        receivers: [],
        contractAddress: ZeroAddress,
        feeController: ZeroAddress,
        externalPriceAddress: ZeroAddress,
        templateAddress: ZeroAddress,
        isCrypto: true,
    };
};
export const getNativeTokenPriceConfig = (amount, receiver) => {
    return getCryptoPriceConfig(amount, receiver, ZeroAddress);
};
export const getExpirableDurationConfig = (durationOfPlan) => {
    return {
        isRedemptionAmountFixed: false,
        redemptionType: PlanRedemptionType.ONLY_SUBSCRIBER,
        proofRequired: false,
        durationSecs: durationOfPlan,
        amount: 1n,
        minAmount: 1n,
        maxAmount: 1n,
    };
};
export const getNonExpirableDurationConfig = () => {
    return getExpirableDurationConfig(0n);
};
export const getFixedCreditsConfig = (creditsGranted, creditsPerRequest = 1n) => {
    return {
        isRedemptionAmountFixed: true,
        redemptionType: PlanRedemptionType.ONLY_SUBSCRIBER,
        proofRequired: false,
        durationSecs: 0n,
        amount: creditsGranted,
        minAmount: creditsPerRequest,
        maxAmount: creditsPerRequest,
    };
};
export const getDynamicCreditsConfig = (creditsGranted, minCreditsPerRequest = 1n, maxCreditsPerRequest = 1n) => {
    return {
        isRedemptionAmountFixed: false,
        redemptionType: PlanRedemptionType.ONLY_SUBSCRIBER,
        proofRequired: false,
        durationSecs: 0n,
        amount: creditsGranted,
        minAmount: minCreditsPerRequest,
        maxAmount: maxCreditsPerRequest,
    };
};
export const setRedemptionType = (creditsConfig, redemptionType) => {
    return {
        ...creditsConfig,
        redemptionType,
    };
};
export const setProofRequired = (creditsConfig, proofRequired = true) => {
    return {
        ...creditsConfig,
        proofRequired,
    };
};
/**
 * Build a pay-as-you-go price configuration.
 *
 * For pay-as-you-go plans, the template address must come from the API deployment info.
 */
export const getPayAsYouGoPriceConfig = (amount, receiver, tokenAddress = ZeroAddress, templateAddress) => {
    if (!isEthereumAddress(receiver))
        throw new Error(`Receiver address ${receiver} is not a valid Ethereum address`);
    if (!templateAddress) {
        throw new Error('templateAddress is required. Use ContractsAPI.getPayAsYouGoTemplateAddress() or Payments.plans.getPayAsYouGoPriceConfig()');
    }
    return {
        tokenAddress,
        amounts: [amount],
        receivers: [receiver],
        contractAddress: ZeroAddress,
        feeController: ZeroAddress,
        externalPriceAddress: ZeroAddress,
        templateAddress,
        isCrypto: true,
    };
};
/**
 * Build a pay-as-you-go credits configuration.
 *
 * Credits are not minted upfront; these values are required for validation only.
 */
export const getPayAsYouGoCreditsConfig = () => {
    return {
        isRedemptionAmountFixed: false,
        redemptionType: PlanRedemptionType.ONLY_SUBSCRIBER,
        proofRequired: false,
        durationSecs: 0n,
        amount: 1n,
        minAmount: 1n,
        maxAmount: 1n,
    };
};
